# Download book and data  


## Download offline handbook  

To download the offline version of this handbook please follow these steps. It will download as an HTML file. You can view the file in your web browser even if you do not have internet access.  

Note: The file is large (>40MB). When you open it, it may take several minutes for the images and the Table of Contents to load/appear. Also, the offline handbook displays as one long page, so please search for specific content using with Ctrl+f (Cmd-f).    

1) Click this link to go [the html file](https://github.com/nsbatra/Epi_R_handbook/tree/master/offline_long) in our Github repository's "offline_long" folder.  

```{r out.height = "50%", out.width= "75%", fig.align = "center", echo=F}
knitr::include_graphics(here::here("images", "download_offline.png"))
```

2) Click the "Download" button. A window will open with HTML source code.  

```{r out.height = "50%", out.width= "75%", fig.align = "center", echo=F}
knitr::include_graphics(here::here("images", "download_offline_button.png"))
```

3) "Save As" the webpage, via right-click (windows) or Cmd-s (mac) - if given the option, set the file type as "Webpage, Complete"  

```{r out.height = "50%", out.width= "75%", fig.align = "center", echo=F}
knitr::include_graphics(here::here("images", "download_offline_saveas.png"))
```





## Download data to follow along  

All the data in this handbook can be downloaded from the **["data" folder](https://github.com/nsbatra/Epi_R_handbook/tree/master/data)** of our Github repository. When you arrive at this folder, click the file you want to download.    

To "follow along" with any of the handbook examples that use the `linelist` dataset, download and load either the "linelist_raw.xlsx" (for the Cleaning page), or "linelist_cleaned.rds" (for any other page).  

Here are detailed instructions for downloading the data files from Github:  




### "Raw" linelist (.xlsx) {-}

Use the "linelist_raw.xlsx" to follow-along with the [Cleaning data and core functions] page. To download the *raw* linelist (uncleaned, with messy data), use the following appraoch.  
*Note: Use "linelist_cleaned.rds" to follow-along with any other page that uses the case linelist.*  

1) Go to the **["data" folder](https://github.com/nsbatra/Epi_R_handbook/tree/master/data)** of our Github repository  
2) Click on "linelist_raw.xlsx"  
3) Click the "Download" button as shown below  
4) Save the file on your computer, and import it into R with the `import()` function from **rio**, as described in the page on [Import and export].  


```{r out.height = "50%", out.width= "75%", fig.align = "center", echo=F}
knitr::include_graphics(here::here("images", "download_xlsx.png"))
```

### Clean linelist (.rds) {-}  

Use the "linelist_cleaned.rds" to follow-along with any page that uses the case linelist (other than the data cleaning page). A .rds file is an R-specific file type that preserves column classes. This ensures you will have minimal cleaning to do after importing.  

1) Go to the **["data" folder](https://github.com/nsbatra/Epi_R_handbook/tree/master/data)** of our Github repository  
2) Click on "linelist_cleaned.rds"  
3) Click the "Download" button as shown below  
4) Save the file on your computer, and import it into R with the `import()` function from **rio**, as described in the page on [Import and export].  



```{r out.height = "50%", out.width= "75%", fig.align = "center", echo=F}
knitr::include_graphics(here::here("images", "download_rds.png"))
```


### .csv files {-} 

It is easy to download a relatively small CSV file directly from Github into R with an R command.  

1) Go to the **["data" folder](https://github.com/nsbatra/Epi_R_handbook/tree/master/data)** of our Github repository  
2) Click on the .csv file of interest  
3) Click on the "Raw" button (you will then see the "raw" csv data, as shown below)  
4) Copy the URL (web address)  
5) Use the URL in the `import()` R command, as shown below  

```{r out.height = "50%", out.width= "75%", fig.align = "center", echo=F}
knitr::include_graphics(here::here("images", "download_csv_raw.png"))
```

Provide the URL to `import()` from **rio** as shown below. 

```{r, eval=F}
# Example of importing the country demographics csv file
country_demographics <- import("https://raw.githubusercontent.com/nsbatra/Epi_R_handbook/master/data/country_demographics.csv")

# Example of importing the likert-scale data csv file
likert_data <- import("https://raw.githubusercontent.com/nsbatra/Epi_R_handbook/master/data/likert_data.csv")
```

If this the above approach not work, you can click the "Raw" button, right-click on the webpage to "Save As" the data as a .txt file to your computer. This file can then be imported into R with `import()`.  


### Shapefiles {-}

Shapefiles have many sub-component files, each with a different file extention. One file will have the ".shp" extension, but others may have ".dbf", ".prj", etc.  

The [GIS basics] page provides links to the *Humanitarian Data Exchange* website where you can download the shapefiles directly. For example, the health facility points can be downloaded [here](https://data.humdata.org/dataset/hotosm_sierra_leone_health_facilities). Download the points shapefile (shp) ".zip" folder. Once saved to your computer, "unzip" the folder. You will see several files with different extensions (e.g. ".shp", ".prj", ".shx") - all these must be saved in the same folder. As described in the [GIS basics] page, provide the filepath and name of the ".shp" file to `st_read()` from the **sf** package.  

Alternatively, you can download the shapefiles from the R Handbook Github "data" folder (see the "shp" sub-folder). However, be aware that you will need to download *each* sub-file individually to your computer. In Github, click on each file individually and download them by clicking on the "Download" button. Below, you can see how the shapefile "sle_adm3" consists of many files - each of which must be downloaded from Github.  

```{r out.height = "50%", fig.align = "center", echo=F}
knitr::include_graphics(here::here("images", "download_shp.png"))
```


### Phylogenetic trees (newick files, saved as .txt)  {-}  

See the page on [Phylogenetic trees]. In Github, navigate to the .txt file, click "Raw", and right-click to download as for the HTML handbook file - save the .txt file on your computer. Then import it with `read.tree()` from the **ape** package.

```{r, eval=F}
ape::read.tree("Shigella_tree.txt")
```

### Other files {-}  

Click on the file in the Gitub "data" folder, and use the "Download" button, as described above for .xlsx files.  




## Background information  

Below are more details about some of the data: 

* The case linelist  
     * A *simulated* Ebola outbreak, expanded by the handbook team from the `ebola_sim` practice dataset in the **outbreaks** package  
     
* Aggregated counts  
     * A simulated dataset of malaria counts by age, day, and facility in a fictional region, which can be downloaded from the data folder noted above as "district_count_data.xlsx"  
     
* Time series and outbreak detection  
     * Campylobacter cases reported in Germany 2002-2011. Available from **surveillance** package
     * Climate data (temperature in degrees celsius and rain fail in millimetres) in Germany 2002-2011. Downloaded from the EU Copernicus satellite reanalysis dataset using the **ecmwfr** package as explained in the time series page.  

* GIS page shapefiles  
     * Downloaded from the Humanitarian Data Exchange (HDX) - see links in the GIS basics page  

* Phylogenetic tree data  
     * Newick file of phylogenetic tree constructed from whole genome sequencing of 299 Shigella sonnei samples and corresponding sample data.
     * The Belgian samples and resulting data are kindly provided by the Belgian NRC for Salmonella and Shigella in the scope of a project conducted by an ECDC EUPHEM Fellow, and will also be published in a manuscript.  
     * The international data are openly available on public databases (ncbi) and have been previously published.  
     

