<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>2 Download handbook and data | The Epidemiologist R Handbook</title>
<meta name="author" content="the handbook team">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.2"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.7/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.5.3/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.5.3/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.2.4/tabs.js"></script><script src="libs/bs3compat-0.2.4/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.15/datatables.js"></script><link href="libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="libs/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet">
<script src="libs/nouislider-7.0.10/jquery.nouislider.min.js"></script><link href="libs/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet">
<script src="libs/selectize-0.12.0/selectize.min.js"></script><link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script><script src="libs/kePrint-0.0.1/kePrint.js"></script><link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet">
<link href="libs/leaflet-1.3.1/leaflet.css" rel="stylesheet">
<script src="libs/leaflet-1.3.1/leaflet.js"></script><link href="libs/leafletfix-1.0.0/leafletfix.css" rel="stylesheet">
<script src="libs/proj4-2.6.2/proj4.min.js"></script><script src="libs/Proj4Leaflet-1.0.1/proj4leaflet.js"></script><link href="libs/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet">
<script src="libs/leaflet-binding-2.0.4.1/leaflet.js"></script><script src="libs/leaflet-providers-1.9.0/leaflet-providers_1.9.0.js"></script><script src="libs/leaflet-providers-plugin-2.0.4.1/leaflet-providers-plugin.js"></script><script src="libs/viz-1.8.2/viz.js"></script><link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script><script src="libs/d3-4.5.0/d3.min.js"></script><script src="libs/sankey-1/sankey.js"></script><script src="libs/sankeyNetwork-binding-0.4/sankeyNetwork.js"></script><script src="libs/plotly-binding-4.9.3/plotly.js"></script><script src="libs/typedarray-0.1/typedarray.min.js"></script><link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script><link href="libs/vis-4.20.1/vis.css" rel="stylesheet">
<script src="libs/vis-4.20.1/vis.min.js"></script><script src="libs/visNetwork-binding-2.0.9/visNetwork.js"></script><link href="libs/font-awesome-4.7.0/css/font-awesome.min.css" rel="stylesheet">
<script src="libs/plotly-basic-1.57.1/plotly-basic-1.57.1.min.js"></script><script src="libs/plotly-cartesian-1.57.1/plotly-cartesian-1.57.1.min.js"></script><script src="https://cdn.jsdelivr.net/autocomplete.js/0/autocomplete.jquery.min.js"></script><script src="https://cdn.jsdelivr.net/npm/mark.js@8.11.1/dist/mark.min.js"></script><!-- CSS --><link rel="stylesheet" href="style_bs4.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">The Epidemiologist R Handbook</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"></a></li>
<li class="book-part">About this book</li>
<li><a class="" href="editorial-and-technical-notes.html"><span class="header-section-number">1</span> Editorial and technical notes</a></li>
<li><a class="active" href="download-handbook-and-data.html"><span class="header-section-number">2</span> Download handbook and data</a></li>
<li class="book-part">Basics</li>
<li><a class="" href="r-basics.html"><span class="header-section-number">3</span> R Basics</a></li>
<li><a class="" href="transition-to-r.html"><span class="header-section-number">4</span> Transition to R</a></li>
<li><a class="" href="suggested-packages-1.html"><span class="header-section-number">5</span> Suggested packages</a></li>
<li><a class="" href="r-projects.html"><span class="header-section-number">6</span> R projects</a></li>
<li><a class="" href="import-and-export.html"><span class="header-section-number">7</span> Import and export</a></li>
<li class="book-part">Data Management</li>
<li><a class="" href="cleaning-data-and-core-functions.html"><span class="header-section-number">8</span> Cleaning data and core functions</a></li>
<li><a class="" href="working-with-dates.html"><span class="header-section-number">9</span> Working with dates</a></li>
<li><a class="" href="characters-and-strings.html"><span class="header-section-number">10</span> Characters and strings</a></li>
<li><a class="" href="factors.html"><span class="header-section-number">11</span> Factors</a></li>
<li><a class="" href="pivoting-data.html"><span class="header-section-number">12</span> Pivoting data</a></li>
<li><a class="" href="grouping-data.html"><span class="header-section-number">13</span> Grouping data</a></li>
<li><a class="" href="joining-data.html"><span class="header-section-number">14</span> Joining data</a></li>
<li><a class="" href="de-duplication.html"><span class="header-section-number">15</span> De-duplication</a></li>
<li><a class="" href="iteration-loops-and-lists.html"><span class="header-section-number">16</span> Iteration, loops, and lists</a></li>
<li class="book-part">Analysis</li>
<li><a class="" href="descriptive-tables.html"><span class="header-section-number">17</span> Descriptive tables</a></li>
<li><a class="" href="simple-statistical-tests.html"><span class="header-section-number">18</span> Simple statistical tests</a></li>
<li><a class="" href="univariate-and-multivariable-regression.html"><span class="header-section-number">19</span> Univariate and multivariable regression</a></li>
<li><a class="" href="missing-data.html"><span class="header-section-number">20</span> Missing data</a></li>
<li><a class="" href="standardization-1.html"><span class="header-section-number">21</span> Standardization</a></li>
<li><a class="" href="moving-averages.html"><span class="header-section-number">22</span> Moving averages</a></li>
<li><a class="" href="time-series-and-outbreak-detection.html"><span class="header-section-number">23</span> Time series and outbreak detection</a></li>
<li><a class="" href="epidemic-modeling.html"><span class="header-section-number">24</span> Epidemic modeling</a></li>
<li><a class="" href="contact-tracing-1.html"><span class="header-section-number">25</span> Contact tracing</a></li>
<li><a class="" href="survey-analysis.html"><span class="header-section-number">26</span> Survey analysis</a></li>
<li><a class="" href="survival-analysis.html"><span class="header-section-number">27</span> Survival analysis</a></li>
<li><a class="" href="gis-basics.html"><span class="header-section-number">28</span> GIS basics</a></li>
<li class="book-part">Data Visualization</li>
<li><a class="" href="tables-for-presentation.html"><span class="header-section-number">29</span> Tables for presentation</a></li>
<li><a class="" href="ggplot-basics.html"><span class="header-section-number">30</span> ggplot basics</a></li>
<li><a class="" href="plot-continuous-data.html"><span class="header-section-number">31</span> Plot continuous data</a></li>
<li><a class="" href="plot-categorical-data.html"><span class="header-section-number">32</span> Plot categorical data</a></li>
<li><a class="" href="ggplot-tips.html"><span class="header-section-number">33</span> ggplot tips</a></li>
<li><a class="" href="epidemic-curves.html"><span class="header-section-number">34</span> Epidemic curves</a></li>
<li><a class="" href="demographic-pyramids-and-likert-scales.html"><span class="header-section-number">35</span> Demographic pyramids and Likert-scales</a></li>
<li><a class="" href="diagrams-and-charts.html"><span class="header-section-number">36</span> Diagrams and charts</a></li>
<li><a class="" href="combinations-analysis.html"><span class="header-section-number">37</span> Combinations analysis</a></li>
<li><a class="" href="heat-plots.html"><span class="header-section-number">38</span> Heat plots</a></li>
<li><a class="" href="transmission-chains.html"><span class="header-section-number">39</span> Transmission chains</a></li>
<li><a class="" href="phylogenetic-trees-1.html"><span class="header-section-number">40</span> Phylogenetic trees</a></li>
<li><a class="" href="interactive-plots.html"><span class="header-section-number">41</span> Interactive plots</a></li>
<li class="book-part">Reports and dashboards</li>
<li><a class="" href="reports-with-r-markdown.html"><span class="header-section-number">42</span> Reports with R Markdown</a></li>
<li><a class="" href="organizing-routine-reports.html"><span class="header-section-number">43</span> Organizing routine reports</a></li>
<li><a class="" href="dashboards-with-r-markdown.html"><span class="header-section-number">44</span> Dashboards with R Markdown</a></li>
<li><a class="" href="dashboards-with-shiny.html"><span class="header-section-number">45</span> Dashboards with Shiny</a></li>
<li class="book-part">Miscellaneous</li>
<li><a class="" href="writing-functions-1.html"><span class="header-section-number">46</span> Writing functions</a></li>
<li><a class="" href="directory-interactions.html"><span class="header-section-number">47</span> Directory interactions</a></li>
<li><a class="" href="version-control-and-collaboration-with-github.html"><span class="header-section-number">48</span> Version control and collaboration with Github</a></li>
<li><a class="" href="common-errors.html"><span class="header-section-number">49</span> Common errors</a></li>
<li><a class="" href="getting-help.html"><span class="header-section-number">50</span> Getting help</a></li>
<li><a class="" href="r-on-network-drives.html"><span class="header-section-number">51</span> R on network drives</a></li>
<li><a class="" href="data-table.html"><span class="header-section-number">52</span> Data Table</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="download-handbook-and-data" class="section level1" number="2">
<h1>
<span class="header-section-number">2</span> Download handbook and data<a class="anchor" aria-label="anchor" href="#download-handbook-and-data"><i class="fas fa-link"></i></a>
</h1>
<div id="download-offline-handbook" class="section level2" number="2.1">
<h2>
<span class="header-section-number">2.1</span> Download offline handbook<a class="anchor" aria-label="anchor" href="#download-offline-handbook"><i class="fas fa-link"></i></a>
</h2>
<p>You can download the offline version of this handbook as an HTML file. Once downloaded, you can view the file in your web browser even if you no longer have internet access. If you are considering offline use of the Epi R Handbook here are a few things to consider:</p>
<ul>
<li>When you open the file it may take a minute or two for the images and the Table of Contents to load<br>
</li>
<li>The offline handbook has a slightly different layout - one very long page with Table of Contents on the left. To search for specific terms use Ctrl+f (Cmd-f)<br>
</li>
<li>See the <a href="suggested-packages-1.html#suggested-packages-1">Suggested packages</a> page to assist you with installing appropriate R packages before you lose internet connectivity<br>
</li>
<li>Install our R package <strong>handbookfiles</strong> that contains all the example data (install process described below)</li>
</ul>
<p><strong>There are two ways you can download the handbook:</strong></p>
<div id="option-1-use-download-link" class="section level3 unnumbered">
<h3>Option 1: Use download link<a class="anchor" aria-label="anchor" href="#option-1-use-download-link"><i class="fas fa-link"></i></a>
</h3>
<p>For quick access, right-click <a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/offline_long/Epi_R_Handbook_offline.html">this link</a> and select “Save link as”.</p>
<p>If on a Mac, use Cmd+click. If on a mobile, press and hold the link and select “Save link”. The handbook will download to your device. If a screen with raw HTML code appears, ensure you have followed the above instructions or try Option 2.</p>
</div>
<div id="option-2-use-our-r-package" class="section level3 unnumbered">
<h3>Option 2: Use our R package<a class="anchor" aria-label="anchor" href="#option-2-use-our-r-package"><i class="fas fa-link"></i></a>
</h3>
<p>We offer an R package called <strong>handbookfiles</strong>. It includes a function <code>download_book()</code> that will download the handbook file from our Github repository to your computer.</p>
<p>This package also contains a function <code>get_data()</code> that will download all the example data to your computer.</p>
<p>To install our R package <strong>handbookfiles</strong>, run the following code. This package is not on CRAN, so use the function <code>p_install_gh()</code> to install it from our <em>epirhandbook</em> Github page.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install the latest version of the Epi R Handbook package</span>
<span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_install_gh.html">p_install_gh</a></span><span class="op">(</span><span class="st">"epirhandbook/handbookfiles"</span><span class="op">)</span></code></pre></div>
<p>Next, run the function <code>download_book()</code> (with empty parentheses) to download the handbook to your computer. A window will appear allowing you to select a save location.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># download the offline handbook to your computer</span>
<span class="fu">download_book</span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="download-data-to-follow-along" class="section level2" number="2.2">
<h2>
<span class="header-section-number">2.2</span> Download data to follow along<a class="anchor" aria-label="anchor" href="#download-data-to-follow-along"><i class="fas fa-link"></i></a>
</h2>
<p>To “follow along” with the handbook pages, you can download the data and use them yourself.</p>
<div id="option-1-use-our-r-package" class="section level3 unnumbered">
<h3>Option 1: Use our R package<a class="anchor" aria-label="anchor" href="#option-1-use-our-r-package"><i class="fas fa-link"></i></a>
</h3>
<p>The easiest approach to download all the data is to install our R package <strong>handbookfiles</strong>. It contains a function <code>get_data()</code> that will save all the example data to a folder of your choice on your computer.</p>
<p>To install our R package <strong>handbookfiles</strong>, run the following code. This package is not on CRAN, so use the function <code>p_install_gh()</code> to install it from our <em>epirhandbook</em> Github page.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install the latest version of the Epi R Handbook package</span>
<span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_install_gh.html">p_install_gh</a></span><span class="op">(</span><span class="st">"epirhandbook/handbookfiles"</span><span class="op">)</span></code></pre></div>
<p>Next, use the function <code>get_data()</code> to download the example data to your computer. Run <code>get_data("all")</code> to get <em>all</em> the example data, or provide a specific file name and extension within the quotes to retrieve only one file.</p>
<p>The data have already been downloaded with the package, and simply need to be transferred out to a folder on your computer. A pop-up window will appear, allowing you to select a save folder location. We suggest you create a new “data” folder as there are about 30 files (including example data and example outputs).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># download all the example data into a folder on your computer</span>
<span class="fu">get_data</span><span class="op">(</span><span class="st">"all"</span><span class="op">)</span></code></pre></div>
<p>If you wish, you can also review all the data in this handbook in the <strong><a href="https://github.com/epirhandbook/Epi_R_handbook/tree/master/data">“data” folder</a></strong> of our Github repository.</p>
</div>
<div id="option-2-download-one-by-one" class="section level3 unnumbered">
<h3>Option 2: Download one-by-one<a class="anchor" aria-label="anchor" href="#option-2-download-one-by-one"><i class="fas fa-link"></i></a>
</h3>
<p>This option involves downloading the data file-by-file from our Github repository via either a link or an R command specific to the file. Some file types allow a download button, while others can be downloaded via an R command.</p>
<div id="case-linelist" class="section level4 unnumbered">
<h4>Case linelist<a class="anchor" aria-label="anchor" href="#case-linelist"><i class="fas fa-link"></i></a>
</h4>
<p>This is a fictional Ebola outbreak, expanded by the handbook team from the <code>ebola_sim</code> practice dataset in the <strong>outbreaks</strong> package.</p>
<ul>
<li><p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/case_linelists/linelist_raw.xlsx" class="download-button">Click to download the “raw” linelist (.xlsx)</a>. The “raw” case linelist is an Excel spreadsheet with messy data. Use this to follow-along with the <a href="cleaning-data-and-core-functions.html#cleaning-data-and-core-functions">Cleaning data and core functions</a> page.</p></li>
<li><p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/case_linelists/linelist_cleaned.rds" class="download-button">Click to download the “clean” linelist (.rds)</a>. Use this file for all other pages of this handbook that use the linelist. A .rds file is an R-specific file type that preserves column classes. This ensures you will have only minimal cleaning to do after importing the data into R.</p></li>
</ul>
<p><em>Other related files:</em></p>
<ul>
<li><p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/case_linelists/linelist_cleaned.xlsx" class="download-button">Click to download the “clean” linelist as an Excel file</a></p></li>
<li><p>Part of the cleaning page uses a “cleaning dictionary” (.csv file). You can load it directly into R by running the following commands in R.</p></li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load</a></span><span class="op">(</span><span class="va">rio</span><span class="op">)</span> <span class="co"># install/load the rio package</span>

<span class="co"># import the file directly from Github</span>
<span class="va">cleaning_dict</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">"https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/case_linelists/cleaning_dict.csv"</span><span class="op">)</span></code></pre></div>
</div>
<div id="data_malaria" class="section level4 unnumbered">
<h4>Malaria count data<a class="anchor" aria-label="anchor" href="#data_malaria"><i class="fas fa-link"></i></a>
</h4>
<p>These data are fictional counts of malaria cases by age group, facility, and day. A .rds file is an R-specific file type that preserves column classes. This ensures you will have only minimal cleaning to do after importing the data into R.</p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/malaria_facility_count_data.rds" class="download-button">
Click to download
<span>the malaria count data (.rds file)</span>
</a></p>
</div>
<div id="likert-scale-data" class="section level4 unnumbered">
<h4>Likert-scale data<a class="anchor" aria-label="anchor" href="#likert-scale-data"><i class="fas fa-link"></i></a>
</h4>
<p>These are fictional data from a Likert-style survey, used in the page on <a href="demographic-pyramids-and-likert-scales.html#demographic-pyramids-and-likert-scales">Demographic pyramids and Likert-scales</a>. You can load these data directly into R by running the following commands in R.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load</a></span><span class="op">(</span><span class="va">rio</span><span class="op">)</span> <span class="co"># install/load the rio package</span>

<span class="co"># import the file directly from Github</span>
<span class="va">likert_data</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/nsbatra/Epi_R_handbook/master/data/likert_data.csv"</span><span class="op">)</span></code></pre></div>
</div>
<div id="flexdashboard" class="section level4 unnumbered">
<h4>Flexdashboard<a class="anchor" aria-label="anchor" href="#flexdashboard"><i class="fas fa-link"></i></a>
</h4>
<p>Below are links to the file associated with the page on <a href="dashboards-with-r-markdown.html#dashboards-with-r-markdown">Dashboards with R Markdown</a>:</p>
<ul>
<li>To download the R Markdown for the outbreak dashboard, right-click this <a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/flexdashboard/outbreak_dashboard.Rmd">link</a> (Cmd+click for Mac) and select “Save link as”.<br>
</li>
<li>To download the HTML dashboard, right-click this <a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/flexdashboard/outbreak_dashboard_test.html">link</a> (Cmd+click for Mac) and select “Save link as”.</li>
</ul>
</div>
</div>
<div id="contact-tracing" class="section level3 unnumbered">
<h3>Contact Tracing<a class="anchor" aria-label="anchor" href="#contact-tracing"><i class="fas fa-link"></i></a>
</h3>
<p>The <a href="contact-tracing-1.html#contact-tracing-1">Contact Tracing</a> page demonstrated analysis of contact tracing data, using example data from <a href="https://github.com/WorldHealthOrganization/godata/tree/master/analytics/r-reporting">Go.Data</a>. The data used in the page can be downloaded as .rds files by clicking the following links:</p>
<p><a href="https://github.com/WorldHealthOrganization/godata/blob/master/analytics/r-reporting/data/cases_clean.rds" class="download-button">
Click to download
<span>the case investigation data (.rds file)</span>
</a></p>
<p><a href="https://github.com/WorldHealthOrganization/godata/blob/master/analytics/r-reporting/data/contacts_clean.rds" class="download-button">
Click to download
<span>the contact registration data (.rds file)</span>
</a></p>
<p><a href="https://github.com/WorldHealthOrganization/godata/blob/master/analytics/r-reporting/data/followups_clean.rds" class="download-button">
Click to download
<span>the contact follow-up data (.rds file)</span>
</a></p>
<p><span style="color: black;"><strong><em>NOTE:</em></strong> Structured contact tracing data from other software (e.g. KoBo, DHIS2 Tracker, CommCare) may look different. If you would like to contribute alternative sample data or content for this page, please <a href="index.html#contact_us">contact us</a>.</span></p>
<p><span style="color: darkgreen;"><strong><em>TIP:</em></strong> If you are deploying Go.Data and want to connect to your instance’s API, see the Import and export page <a href="import-and-export.html#import_api">(API section)</a> and the <a href="https://community-godata.who.int/">Go.Data Community of Practice</a>.</span></p>
<div id="gis" class="section level4 unnumbered">
<h4>GIS<a class="anchor" aria-label="anchor" href="#gis"><i class="fas fa-link"></i></a>
</h4>
<p>Shapefiles have many sub-component files, each with a different file extention. One file will have the “.shp” extension, but others may have “.dbf”, “.prj”, etc.</p>
<p>The <a href="gis-basics.html#gis-basics">GIS basics</a> page provides links to the <em>Humanitarian Data Exchange</em> website where you can download the shapefiles directly as zipped files.</p>
<p>For example, the health facility points data can be downloaded <a href="https://data.humdata.org/dataset/hotosm_sierra_leone_health_facilities">here</a>. Download “hotosm_sierra_leone_health_facilities_points_shp.zip”. Once saved to your computer, “unzip” the folder. You will see several files with different extensions (e.g. “.shp”, “.prj”, “.shx”) - all these must be saved to the same folder on your computer. Then to import into R, provide the file path and name of the “.shp” file to <code>st_read()</code> from the <strong>sf</strong> package (as described in the <a href="gis-basics.html#gis-basics">GIS basics</a> page).</p>
<p>If you follow Option 1 to download all the example data (via our R package <strong>handbookfiles</strong>), all the shapefiles are included.</p>
<p>Alternatively, you can download the shapefiles from the R Handbook Github “data” folder (see the “gis” sub-folder). However, be aware that you will need to download <em>each</em> sub-file individually to your computer. In Github, click on each file individually and download them by clicking on the “Download” button. Below, you can see how the shapefile “sle_adm3” consists of many files - each of which would need to be downloaded from Github.</p>
<div class="inline-figure"><img src="images/download_shp.png" width="918" height="50%" style="display: block; margin: auto;"></div>
</div>
<div id="phylogenetic-trees" class="section level4 unnumbered">
<h4>Phylogenetic trees<a class="anchor" aria-label="anchor" href="#phylogenetic-trees"><i class="fas fa-link"></i></a>
</h4>
<p>See the page on <a href="phylogenetic-trees-1.html#phylogenetic-trees-1">Phylogenetic trees</a>. Newick file of phylogenetic tree constructed from whole genome sequencing of 299 Shigella sonnei samples and corresponding sample data (converted to a text file). The Belgian samples and resulting data are kindly provided by the Belgian NRC for Salmonella and Shigella in the scope of a project conducted by an ECDC EUPHEM Fellow, and will also be published in a manuscript. The international data are openly available on public databases (ncbi) and have been previously published.</p>
<ul>
<li>To download the “Shigella_tree.txt” phylogenetic tree file, right-click this <a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/phylo/Shigella_tree.txt">link</a> (Cmd+click for Mac) and select “Save link as”.<br>
</li>
<li>To download the “sample_data_Shigella_tree.csv” with additional information on each sample, right-click this <a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/phylo/sample_data_Shigella_tree.csv">link</a> (Cmd+click for Mac) and select “Save link as”.<br>
</li>
<li>To see the new, created subset-tree, right-click this <a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/phylo/Shigella_subtree_2.txt">link</a> (Cmd+click for Mac) and select “Save link as”. The .txt file will download to your computer.</li>
</ul>
<p>You can then import the .txt files with <code>read.tree()</code> from the <strong>ape</strong> package, as explained in the page.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span><span class="st">"Shigella_tree.txt"</span><span class="op">)</span></code></pre></div>
</div>
<div id="standardization" class="section level4 unnumbered">
<h4>Standardization<a class="anchor" aria-label="anchor" href="#standardization"><i class="fas fa-link"></i></a>
</h4>
<p>See the page on <a href="standardization-1.html#standardization-1">Standardization</a>. You can load the data directly from our Github repository on the internet into your R session with the following commands:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install/load the rio package</span>
<span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load</a></span><span class="op">(</span><span class="va">rio</span><span class="op">)</span> 

<span class="co">##############</span>
<span class="co"># Country A</span>
<span class="co">##############</span>
<span class="co"># import demographics for country A directly from Github</span>
<span class="va">countryAdemo</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">"https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/standardization/country_demographics.csv"</span><span class="op">)</span>

<span class="co"># import deaths for country A directly from Github</span>
<span class="va">A_deaths</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">"https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/standardization/deaths_countryA.csv"</span><span class="op">)</span>

<span class="co">##############</span>
<span class="co"># Country B</span>
<span class="co">##############</span>
<span class="co"># import demographics for country B directly from Github</span>
<span class="va">countryBdemo</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">"https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/standardization/country_demographics_2.csv"</span><span class="op">)</span>

<span class="co"># import deaths for country B directly from Github</span>
<span class="va">B_deaths</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">"https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/standardization/deaths_countryB.csv"</span><span class="op">)</span>


<span class="co">###############</span>
<span class="co"># Reference Pop</span>
<span class="co">###############</span>
<span class="co"># import demographics for country B directly from Github</span>
<span class="va">standard_pop_data</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">"https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/standardization/world_standard_population_by_sex.csv"</span><span class="op">)</span></code></pre></div>
</div>
<div id="data_outbreak" class="section level4 unnumbered">
<h4>Time series and outbreak detection<a class="anchor" aria-label="anchor" href="#data_outbreak"><i class="fas fa-link"></i></a>
</h4>
<p>See the page on <a href="time-series-and-outbreak-detection.html#time-series-and-outbreak-detection">Time series and outbreak detection</a>. We use campylobacter cases reported in Germany 2002-2011, as available from the <strong>surveillance</strong> R package.</p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/campylobacter_germany.xlsx" class="download-button">
Click to download
<span> Campylobacter in Germany (.xlsx)</span>
</a></p>
<p>Also used are climate data (temperature in degrees celsius and rain fail in millimetres) in Germany 2002-2011. These were downloaded from the EU Copernicus satellite reanalysis dataset using the <strong>ecmwfr</strong> package. You will need to download all of these and import them with <code><a href="https://r-spatial.github.io/stars/reference/read_stars.html">stars::read_stars()</a></code> as explained in the time series page.</p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2002.nc" class="download-button">
Click to download
<span> Germany weather 2002 (.nc file)</span>
</a></p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2003.nc" class="download-button">
Click to download
<span> Germany weather 2003 (.nc file)</span>
</a></p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2004.nc" class="download-button">
Click to download
<span> Germany weather 2004 (.nc file)</span>
</a></p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2005.nc" class="download-button">
Click to download
<span> Germany weather 2005 (.nc file)</span>
</a></p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2006.nc" class="download-button">
Click to download
<span> Germany weather 2006 (.nc file)</span>
</a></p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2007.nc" class="download-button">
Click to download
<span> Germany weather 2007 (.nc file)</span>
</a></p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2008.nc" class="download-button">
Click to download
<span> Germany weather 2008 (.nc file)</span>
</a></p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2009.nc" class="download-button">
Click to download
<span> Germany weather 2009 (.nc file)</span>
</a></p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2010.nc" class="download-button">
Click to download
<span> Germany weather 2010 (.nc file)</span>
</a></p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/time_series/weather/germany_weather2011.nc" class="download-button">
Click to download
<span> Germany weather 2011 (.nc file)</span>
</a></p>
</div>
<div id="data_survey" class="section level4 unnumbered">
<h4>Survey analysis<a class="anchor" aria-label="anchor" href="#data_survey"><i class="fas fa-link"></i></a>
</h4>
<p>For the <a href="https://epirhandbook.com/survey-analysis.html">survey analysis</a> page we use fictional mortality survey data based off MSF OCA survey templates,
this fictional data was generated as part of the <a href="https://r4epis.netlify.app/">“R4Epis” project</a>.</p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/surveys/survey_data.xlsx" class="download-button">
Click to download
<span> Fictional survey data (.xlsx)</span>
</a></p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/surveys/survey_dict.xlsx" class="download-button">
Click to download
<span> Fictional survey data dictionary (.xlsx)</span>
</a></p>
<p><a href="https://github.com/epirhandbook/Epi_R_handbook/raw/master/data/surveys/population.xlsx" class="download-button">
Click to download
<span> Fictional survey population data (.xlsx)</span>
</a></p>
</div>
<div id="data_shiny" class="section level4 unnumbered">
<h4>Shiny<a class="anchor" aria-label="anchor" href="#data_shiny"><i class="fas fa-link"></i></a>
</h4>
<p>The page on <a href="dashboards-with-shiny.html#dashboards-with-shiny">Dashboards with Shiny</a> demonstrated constructions of a simple app to display malaria data.</p>
<p>To download the R files that produce the Shiny app:</p>
<p>You can <a href="https://github.com/epirhandbook/Epi_R_handbook/blob/master/data/malaria_app/app.R" class="download-button">
click here to download the app.R file<span> that contains both the UI and Server code for the Shiny app.</span></a></p>
<p>You can <a href="https://github.com/epirhandbook/Epi_R_handbook/blob/master/data/malaria_app/data/facility_count_data.rds" class="download-button">
click here to download the facility_count_data.rds file<span> that contains malaria data for the Shiny app. Note that you may need to store it within a “data” folder for the here() file paths to work correctly.</span></a></p>
<p>You can <a href="https://github.com/epirhandbook/Epi_R_handbook/blob/master/data/malaria_app/global.R" class="download-button">
click here to download the global.R file<span> that should run prior to the app opening, as explained in the page.</span></a></p>
<p>You can <a href="https://github.com/epirhandbook/Epi_R_handbook/blob/master/data/malaria_app/funcs/plot_epicurve.R" class="download-button">
click here to download the plot_epicurve.R file<span> that is sourced by global.R. Note that you may need to store it within a “funcs” folder for the here() file paths to work correctly.</span></a></p>

</div>
</div>
</div>
</div>



<div class="chapter-nav">
<div class="prev"><a href="editorial-and-technical-notes.html"><span class="header-section-number">1</span> Editorial and technical notes</a></div>
<div class="next"><a href="r-basics.html"><span class="header-section-number">3</span> R Basics</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#download-handbook-and-data"><span class="header-section-number">2</span> Download handbook and data</a></li>
<li>
<a class="nav-link" href="#download-offline-handbook"><span class="header-section-number">2.1</span> Download offline handbook</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#option-1-use-download-link">Option 1: Use download link</a></li>
<li><a class="nav-link" href="#option-2-use-our-r-package">Option 2: Use our R package</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#download-data-to-follow-along"><span class="header-section-number">2.2</span> Download data to follow along</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#option-1-use-our-r-package">Option 1: Use our R package</a></li>
<li><a class="nav-link" href="#option-2-download-one-by-one">Option 2: Download one-by-one</a></li>
<li><a class="nav-link" href="#contact-tracing">Contact Tracing</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>
</div>
  

  

</div>
 <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>The Epidemiologist R Handbook</strong>" was written by the handbook team. It was last built on 2021-05-06.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>
</html>
