<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <title>38 Phylogenetic trees | The Epidemiologist R Handbook</title>

    <meta name="author" content="the handbook team" />
  
  <!-- JS -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.2"></script>
  <script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script>
    <script src="libs/header-attrs-2.8/header-attrs.js"></script>
    <script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
    <link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet" />
    <script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script>
    <script src="libs/bs3compat-0.2.5.1/tabs.js"></script>
    <script src="libs/bs3compat-0.2.5.1/bs3compat.js"></script>
    <link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet" />
    <script src="libs/bs4_book-1.0.0/bs4_book.js"></script>
    <script src="https://cdn.jsdelivr.net/autocomplete.js/0/autocomplete.jquery.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/mark.js@8.11.1/dist/mark.min.js"></script>

  <!-- CSS -->
    <link rel="stylesheet" href="style_bs4.css" />
    
</head>

<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book">
    <a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">The Epidemiologist R Handbook</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
      </form>

      <nav aria-label="Table of contents">
        <h2>Table of contents</h2>
        <div id="book-toc"></div>

        <div class="book-extra">
          <p><a id="book-repo" href="#">View book source <i class="fab fa-github"></i></a></li></p>
        </div>
      </nav>
    </div>
  </header>

  <main class="col-sm-12 col-md-9 col-lg-7" id="content">
<div id="phylogenetic-trees-1" class="section level1" number="38">
<h1><span class="header-section-number">38</span> Phylogenetic trees</h1>
<!-- ======================================================= -->
<div id="overview-8" class="section level2" number="38.1">
<h2><span class="header-section-number">38.1</span> Overview</h2>
<p><strong>Phylogenetic trees</strong> are used to visualize and describe the relatedness and evolution of organisms based on the sequence of their genetic code.</p>
<p>They can be constructed from genetic sequences using distance-based methods (such as neighbor-joining method) or character-based methods (such as maximum likelihood and Bayesian Markov Chain Monte Carlo method). Next-generation sequencing (NGS) has become more affordable and is becoming more widely used in public health to describe pathogens causing infectious diseases. Portable sequencing devices decrease the turn around time and hold promises to make data available for the support of outbreak investigation in real-time. NGS data can be used to identify the origin or source of an outbreak strain and its propagation, as well as determine presence of antimicrobial resistance genes. To visualize the genetic relatedness between samples a phylogenetic tree is constructed.</p>
<p>In this page we will learn how to use the <strong>ggtree</strong> package, which allows for combined visualization of phylogenetic trees with additional sample data in form of a dataframe. This will enable us to observe patterns and improve understanding of the outbreak dynamic.</p>
<!-- ======================================================= -->
</div>
<div id="preparation-31" class="section level2" number="38.2">
<h2><span class="header-section-number">38.2</span> Preparation</h2>
<div id="load-packages-28" class="section level3 unnumbered">
<h3>Load packages</h3>
<p>This code chunk shows the loading of required packages. In this handbook we emphasize <code>p_load()</code> from <strong>pacman</strong>, which installs the package if necessary <em>and</em> loads it for use. You can also load installed packages with <code>library()</code> from <strong>base</strong> R. See the page on <a href="r-basics.html#r-basics">R basics</a> for more information on R packages.</p>
</div>
<div id="import-data-24" class="section level3 unnumbered">
<h3>Import data</h3>
<p>The data for this page can be downloaded with the instructions on the <a href="download-handbook-and-data.html#download-handbook-and-data">Download handbook and data</a> page.</p>
<p>There are several different formats in which a phylogenetic tree can be stored (eg. Newick, NEXUS, Phylip). A common one is the Newick file format (.nwk), which is the standard for representing trees in computer-readable form. This means an entire tree can be expressed in a string format such as “((t2:0.04,t1:0.34):0.89,(t5:0.37,(t4:0.03,t3:0.67):0.9):0.59);”, listing all nodes and tips and their relationship (branch length) to each other.</p>
<p>Note: It is important to understand that the phylogenetic tree file in itself does not contain sequencing data, but is merely the result of the genetic distances between the sequences. We therefore cannot extract sequencing data from a tree file.</p>
<p>First, we use the <code>read.tree()</code> function from <strong>ape</strong> package to import a Newick phylogenetic tree file in .txt format, and store it in a list object of class “phylo”. If necessary, use the <code>here()</code> function from the <strong>here</strong> package to specify the relative file path.</p>
<p>Note: In this case the newick tree is saved as a .txt file for easier handling and downloading from Github.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="phylogenetic-trees-1.html#cb8-1" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">read.tree</span>(<span class="st">&quot;Shigella_tree.txt&quot;</span>)</span></code></pre></div>
<p>We inspect our tree object and see it contains 299 tips (or samples) and 236 nodes.</p>
<p>Second, we import a table stored as a .csv file with additional information for each sequenced sample, such as gender, country of origin and attributes for antimicrobial resistance, using the <code>import()</code> function from the <strong>rio</strong> package:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="phylogenetic-trees-1.html#cb9-1" aria-hidden="true" tabindex="-1"></a>sample_data <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">&quot;sample_data_Shigella_tree.csv&quot;</span>)</span></code></pre></div>
<p>Below are the first 50 rows of the data:</p>
</div>
<div id="clean-and-inspect" class="section level3 unnumbered">
<h3>Clean and inspect</h3>
<p>We clean and inspect our data: In order to assign the correct sample data to the phylogenetic tree, the values in the column <code>Sample_ID</code> in the <code>sample_data</code> data frame need to match the <code>tip.labels</code> values in the <code>tree</code> file:</p>
<p>We check the formatting of the <code>tip.labels</code> in the <code>tree</code> file by looking at the first 6 entries using with <code>head()</code> from <strong>base</strong> R.</p>
<p>We also make sure the first column in our <code>sample_data</code> data frame is <code>Sample_ID</code>. We look at the column names of our dataframe using <code>colnames()</code> from <strong>base</strong> R.</p>
<p>We look at the <code>Sample_IDs</code> in the data frame to make sure the formatting is the same than in the <code>tip.label</code> (eg. letters are all capitals, no extra underscores <code>_</code> between letters and numbers, etc.)</p>
<p>We can also compare if all samples are present in the <code>tree</code> file and vice versa by generating a logical vector of TRUE or FALSE where they do or do not match. These are not printed here, for simplicity.</p>
<p>We can use these vectors to show any sample IDs that are not on the tree (there are none).</p>
<p>Upon inspection we can see that the format of <code>Sample_ID</code> in the dataframe corresponds to the format of sample names at the <code>tip.labels</code>. These do not have to be sorted in the same order to be matched.</p>
<p>We are ready to go!</p>
<!-- ======================================================= -->
</div>
</div>
<div id="simple-tree-visualization" class="section level2" number="38.3">
<h2><span class="header-section-number">38.3</span> Simple tree visualization</h2>
<div id="different-tree-layouts" class="section level3 unnumbered">
<h3>Different tree layouts</h3>
<p><strong>ggtree</strong> offers many different layout formats and some may be more suitable for your specific purpose than others. Below are a few demonstrations. For other options see this <a href="http://yulab-smu.top/treedata-book/chapter4.html">online book</a>.</p>
<p>Here are some example tree layouts:</p>
</div>
<div id="simple-tree-plus-sample-data" class="section level3 unnumbered">
<h3>Simple tree plus sample data</h3>
<p>The <strong>%&lt;+%</strong> operator is used to connect the <code>sample_data</code> data frame to the <code>tree</code> file.
The most easy annotation of your tree is the addition of the sample names at the tips, as well as coloring of tip points and if desired the branches:</p>
<p>Here is an example of a circular tree:</p>
<p>You can export your tree plot with <code>ggsave()</code> as you would any other ggplot object. Written this way, <code>ggsave()</code> saves the last image produced to the file path you specify. Remember that you can use <code>here()</code> and relative file paths to easily save in subfolders, etc.</p>
<!-- ======================================================= -->
</div>
</div>
<div id="tree-manipulation" class="section level2" number="38.4">
<h2><span class="header-section-number">38.4</span> Tree manipulation</h2>
<p>Sometimes you may have a very large phylogenetic tree and you are only interested in one part of the tree. For example, if you produced a tree including historical or international samples to get a large overview of where your dataset might fit in the bigger picture. But then to look closer at your data you want to inspect only that portion of the bigger tree.</p>
<p>Since the phylogenetic tree file is just the output of sequencing data analysis, we can not manipulate the order of the nodes and branches in the file itself. These have already been determined in previous analysis from the raw NGS data. We are able though to zoom into parts, hide parts and even subset part of the tree.</p>
<div id="zoom-in" class="section level3 unnumbered">
<h3>Zoom in</h3>
<p>If you don’t want to “cut” your tree, but only inspect part of it more closely you can zoom in to view a specific part.</p>
<p>First, we plot the entire tree in linear format and add numeric labels to each node in the tree.</p>
<p>To zoom in to one particular branch (sticking out to the right), use <code>viewClade()</code> on the ggtree object <code>p</code> and provide the node number to get a closer look:</p>
</div>
<div id="collapsing-branches" class="section level3 unnumbered">
<h3>Collapsing branches</h3>
<p>However, we may want to ignore this branch and can collapse it at that same node (node nr. 452) using <code>collapse()</code>. This tree is defined as <code>p_collapsed</code>.</p>
<p>For clarity, when we print <code>p_collapsed</code>, we add a <code>geom_point2()</code> (a blue diamond) at the node of the collapsed branch.</p>
</div>
<div id="subsetting-a-tree" class="section level3 unnumbered">
<h3>Subsetting a tree</h3>
<p>If we want to make a more permanent change and create a new, reduced tree to work with we can subset part of it with <code>tree_subset()</code>. Then you can save it as new newick tree file or .txt file.</p>
<p>First, we inspect the tree nodes and tip labels in order to decide what to subset.</p>
<p>Now, say we have decided to subset the tree at node 528 (keep only tips within this branch after node 528) and we save it as a new <code>sub_tree1</code> object:</p>
<p>Lets have a look at the subset tree 1:</p>
<p>You can also subset based on one particular sample, specifying how many nodes “backwards” you want to include. Let’s subset the same part of the tree based on a sample, in this case S17BD07692, going back 9 nodes and we save it as a new <code>sub_tree2</code> object:</p>
<p>Lets have a look at the subset tree 2:</p>
<p>You can also save your new tree either as a Newick type or even a text file using the <code>write.tree()</code> function from <strong>ape</strong> package:</p>
</div>
<div id="rotating-nodes-in-a-tree" class="section level3 unnumbered">
<h3>Rotating nodes in a tree</h3>
<p>As mentioned before we cannot change the order of tips or nodes in the tree, as this is based on their genetic relatedness and is not subject to visual manipulation. But we can rote branches around nodes if that eases our visualization.</p>
<p>First, we plot our new subset tree 2 with node labels to choose the node we want to manipulate and store it an a ggtree plot object <code>p</code>.</p>
<p>We can then manipulate nodes by applying <strong>ggtree::rotate()</strong> or <strong>ggtree::flip()</strong>:
Note: to illustrate which nodes we are manipulating we first apply the <strong>geom_hilight()</strong> function from <strong>ggtree</strong> to highlight the samples in the nodes we are interested in and store that ggtree plot object in a new object <code>p1</code>.</p>
<p>Now we can rotate node 37 in object <code>p1</code> so that the samples on node 38 move to the top. We store the rotated tree in a new object <code>p2</code>.</p>
<p>Or we can use the flip command to rotate node 36 in object <code>p1</code> and switch node 37 to the top and node 39 to the bottom. We store the flipped tree in a new object <code>p3</code>.</p>
</div>
<div id="example-subtree-with-sample-data-annotation" class="section level3 unnumbered">
<h3>Example subtree with sample data annotation</h3>
<p>Lets say we are investigating the cluster of cases with clonal expansion which occurred in 2017 and 2018 at node 39 in our sub-tree. We add the year of strain isolation as well as travel history and color by country to see origin of other closely related strains:</p>
<p>Our observation points towards an import event of strains from Asia, which then circulated in Belgium over the years and seem to have caused our latest outbreak.</p>
<!-- ======================================================= -->
</div>
</div>
<div id="more-complex-trees-adding-heatmaps-of-sample-data" class="section level2 unnumbered">
<h2>More complex trees: adding heatmaps of sample data</h2>
<p>We can add more complex information, such as categorical presence of antimicrobial resistance genes and numeric values for actually measured resistance to antimicrobials in form of a heatmap using the <strong>ggtree::gheatmap()</strong> function.</p>
<p>First we need to plot our tree (this can be either linear or circular) and store it in a new ggtree plot object <code>p</code>: We will use the sub_tree from part 3.)</p>
<p>Second, we prepare our data. To visualize different variables with new color schemes, we subset our dataframe to the desired variable. It is important to add the <code>Sample_ID</code> as rownames otherwise it cannot match the data to the tree <code>tip.labels</code>:</p>
<p>In our example we want to look at gender and mutations that could confer resistance to Ciprofloxacin, an important first line antibiotic used to treat Shigella infections.</p>
<p>We create a dataframe for gender:</p>
<p>We create a dataframe for mutations in the gyrA gene, which confer Ciprofloxacin resistance:</p>
<p>We create a dataframe for the measured minimum inhibitory concentration (MIC) for Ciprofloxacin from the laboratory:</p>
<p>We create a first plot adding a binary heatmap for gender to the phylogenetic tree and storing it in a new ggtree plot object <code>h1</code>:</p>
<p>Then we add information on mutations in the gyrA gene, which confer resistance to Ciprofloxacin:</p>
<p>Note: The presence of chromosomal point mutations in WGS data was prior determined using the PointFinder tool developed by Zankari et al. (see reference in the additional references section)</p>
<p>First, we assign a new color scheme to our existing plot object <code>h1</code> and store it in a now object <code>h2</code>. This enables us to define and change the colors for our second variable in the heatmap.</p>
<p>Then we add the second heatmap layer to <code>h2</code> and store the combined plots in a new object <code>h3</code>:</p>
<p>We repeat the above process, by first adding a new color scale layer to our existing object <code>h3</code>, and then adding the continuous data on the minimum inhibitory concentration (MIC) of Ciprofloxacin for each strain to the resulting object <code>h4</code> to produce the final object <code>h5</code>:</p>
<p>We can do the same exercise for a linear tree:</p>
<p>First we add gender:</p>
<p>Then we add Ciprofloxacin resistance mutations after adding another color scheme layer:</p>
<p>Then we add the minimum inhibitory concentration determined by the laboratory (MIC):</p>
<!-- ======================================================= -->
</div>
<div id="resources-31" class="section level2" number="38.5">
<h2><span class="header-section-number">38.5</span> Resources</h2>
<p><a href="http://hydrodictyon.eeb.uconn.edu/eebedia/index.php/Ggtree#" class="uri">http://hydrodictyon.eeb.uconn.edu/eebedia/index.php/Ggtree#</a> Clade_Colors
<a href="https://bioconductor.riken.jp/packages/3.2/bioc/vignettes/ggtree/inst/doc/treeManipulation.html" class="uri">https://bioconductor.riken.jp/packages/3.2/bioc/vignettes/ggtree/inst/doc/treeManipulation.html</a>
<a href="https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html" class="uri">https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html</a>
<a href="https://bioconductor.riken.jp/packages/3.8/bioc/vignettes/ggtree/inst/doc/treeManipulation.html" class="uri">https://bioconductor.riken.jp/packages/3.8/bioc/vignettes/ggtree/inst/doc/treeManipulation.html</a></p>
<p>Ea Zankari, Rosa Allesøe, Katrine G Joensen, Lina M Cavaco, Ole Lund, Frank M Aarestrup, PointFinder: a novel web tool for WGS-based detection of antimicrobial resistance associated with chromosomal point mutations in bacterial pathogens, Journal of Antimicrobial Chemotherapy, Volume 72, Issue 10, October 2017, Pages 2764–2768, <a href="https://doi.org/10.1093/jac/dkx217" class="uri">https://doi.org/10.1093/jac/dkx217</a></p>

</div>
</div>
  </main>

  <div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page">
      <h2>On this page</h2>
      <div id="book-on-this-page"></div>

      <div class="book-extra">
        <ul class="list-unstyled">
          <li><a id="book-source" href="#">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="#">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
      </div>
    </nav>
  </div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5">
  <div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>The Epidemiologist R Handbook</strong>" was written by the handbook team. It was last built on 2021-05-30.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
