<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>10 Grouping data | R Handbook for Epidemiologists</title>
<meta name="author" content="the handbook team">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.2"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.6/header-attrs.js"></script><script src="libs/jquery-3.5.1/jquery-3.5.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.5.3/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.5.3/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.2.3.9000/tabs.js"></script><script src="libs/bs3compat-0.2.3.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.15/datatables.js"></script><link href="libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="libs/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet">
<script src="libs/nouislider-7.0.10/jquery.nouislider.min.js"></script><link href="libs/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet">
<script src="libs/selectize-0.12.0/selectize.min.js"></script><link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script><link href="libs/leaflet-1.3.1/leaflet.css" rel="stylesheet">
<script src="libs/leaflet-1.3.1/leaflet.js"></script><link href="libs/leafletfix-1.0.0/leafletfix.css" rel="stylesheet">
<script src="libs/proj4-2.6.2/proj4.min.js"></script><script src="libs/Proj4Leaflet-1.0.1/proj4leaflet.js"></script><link href="libs/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet">
<script src="libs/leaflet-binding-2.0.4.1/leaflet.js"></script><script src="libs/leaflet-providers-1.9.0/leaflet-providers_1.9.0.js"></script><script src="libs/leaflet-providers-plugin-2.0.4.1/leaflet-providers-plugin.js"></script><link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet">
<script src="libs/viz-1.8.2/viz.js"></script><link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script><script src="libs/d3-4.5.0/d3.min.js"></script><script src="libs/sankey-1/sankey.js"></script><script src="libs/sankeyNetwork-binding-0.4/sankeyNetwork.js"></script><script src="libs/plotly-binding-4.9.3/plotly.js"></script><script src="libs/typedarray-0.1/typedarray.min.js"></script><link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script><link href="libs/vis-4.20.1/vis.css" rel="stylesheet">
<script src="libs/vis-4.20.1/vis.min.js"></script><script src="libs/visNetwork-binding-2.0.9/visNetwork.js"></script><link href="libs/font-awesome-4.7.0/css/font-awesome.min.css" rel="stylesheet">
<script src="libs/plotly-basic-1.57.1/plotly-basic-1.57.1.min.js"></script><script src="libs/plotly-cartesian-1.57.1/plotly-cartesian-1.57.1.min.js"></script><script src="https://cdn.jsdelivr.net/autocomplete.js/0/autocomplete.jquery.min.js"></script><script src="https://cdn.jsdelivr.net/npm/mark.js@8.11.1/dist/mark.min.js"></script><!-- CSS --><link rel="stylesheet" href="style_bs4.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">R Handbook for Epidemiologists</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">Welcome - THIS IS A DRAFT</a></li>
<li class="book-part">About this book</li>
<li><a class="" href="style-and-editorial-notes.html"><span class="header-section-number">1</span> Style and editorial notes</a></li>
<li><a class="" href="datasets-used.html"><span class="header-section-number">2</span> Datasets used</a></li>
<li class="book-part">Basics</li>
<li><a class="" href="r-basics.html"><span class="header-section-number">3</span> R Basics</a></li>
<li><a class="" href="importing-and-exporting.html"><span class="header-section-number">4</span> Importing and exporting</a></li>
<li><a class="" href="r-projects.html"><span class="header-section-number">5</span> R projects</a></li>
<li><a class="" href="suggested-packages.html"><span class="header-section-number">6</span> Suggested packages</a></li>
<li class="book-part">Data Management</li>
<li><a class="" href="cleaning-data.html"><span class="header-section-number">7</span> Cleaning data</a></li>
<li><a class="" href="working-with-dates.html"><span class="header-section-number">8</span> Working with dates</a></li>
<li><a class="" href="missing-data.html"><span class="header-section-number">9</span> Missing data</a></li>
<li><a class="active" href="grouping-data.html"><span class="header-section-number">10</span> Grouping data</a></li>
<li><a class="" href="joining-data.html"><span class="header-section-number">11</span> Joining data</a></li>
<li><a class="" href="characters-and-strings.html"><span class="header-section-number">12</span> Characters and strings</a></li>
<li><a class="" href="de-duplication.html"><span class="header-section-number">13</span> De-duplication</a></li>
<li><a class="" href="iteration-and-loops.html"><span class="header-section-number">14</span> Iteration and loops</a></li>
<li class="book-part">Analysis</li>
<li><a class="" href="descriptive-analysis.html"><span class="header-section-number">15</span> Descriptive analysis</a></li>
<li><a class="" href="simple-statistical-tests.html"><span class="header-section-number">16</span> Simple statistical tests</a></li>
<li><a class="" href="standardization.html"><span class="header-section-number">17</span> Standardization</a></li>
<li><a class="" href="moving-averages.html"><span class="header-section-number">18</span> Moving averages</a></li>
<li><a class="" href="outbreak-detection.html"><span class="header-section-number">19</span> Outbreak detection</a></li>
<li><a class="" href="time-series-analysis.html"><span class="header-section-number">20</span> Time series analysis</a></li>
<li><a class="" href="epidemic-modeling.html"><span class="header-section-number">21</span> Epidemic modeling</a></li>
<li><a class="" href="modeling.html"><span class="header-section-number">22</span> Modeling</a></li>
<li><a class="" href="survey-analysis.html"><span class="header-section-number">23</span> Survey analysis</a></li>
<li><a class="" href="survival-analysis-1.html"><span class="header-section-number">24</span> Survival analysis</a></li>
<li><a class="" href="gis-basics.html"><span class="header-section-number">25</span> GIS basics</a></li>
<li class="book-part">Data Vizualization</li>
<li><a class="" href="ggplot-tips.html"><span class="header-section-number">26</span> ggplot tips</a></li>
<li><a class="" href="epidemic-curves.html"><span class="header-section-number">27</span> Epidemic curves</a></li>
<li><a class="" href="plot-continuous-data.html"><span class="header-section-number">28</span> Plot continuous data</a></li>
<li><a class="" href="plot-categorical-data.html"><span class="header-section-number">29</span> Plot categorical data</a></li>
<li><a class="" href="tables.html"><span class="header-section-number">30</span> Tables</a></li>
<li><a class="" href="age-pyramids.html"><span class="header-section-number">31</span> Age pyramids</a></li>
<li><a class="" href="diagrams.html"><span class="header-section-number">32</span> Diagrams</a></li>
<li><a class="" href="combination-analysis.html"><span class="header-section-number">33</span> Combination analysis</a></li>
<li><a class="" href="heatmaps-density-plots.html"><span class="header-section-number">34</span> Heatmaps &amp; density plots</a></li>
<li><a class="" href="transmission-chains.html"><span class="header-section-number">35</span> Transmission chains</a></li>
<li><a class="" href="phylogenetic-trees.html"><span class="header-section-number">36</span> Phylogenetic trees</a></li>
<li><a class="" href="interactive-plots.html"><span class="header-section-number">37</span> Interactive plots</a></li>
<li class="book-part">Advanced</li>
<li><a class="" href="errors-warnings-1.html"><span class="header-section-number">38</span> Errors &amp; warnings</a></li>
<li><a class="" href="advanced-rstudio.html"><span class="header-section-number">39</span> Advanced RStudio</a></li>
<li><a class="" href="relational-databases.html"><span class="header-section-number">40</span> Relational databases</a></li>
<li><a class="" href="routine-reports.html"><span class="header-section-number">41</span> Routine reports</a></li>
<li><a class="" href="r-markdown-1.html"><span class="header-section-number">42</span> R Markdown</a></li>
<li><a class="" href="shiny-and-dashboards.html"><span class="header-section-number">43</span> Shiny and dashboards</a></li>
<li><a class="" href="collaboration.html"><span class="header-section-number">44</span> Collaboration</a></li>
<li><a class="" href="writing-functions.html"><span class="header-section-number">45</span> Writing functions</a></li>
<li><a class="" href="r-on-network-drives.html"><span class="header-section-number">46</span> R on network drives</a></li>
<li><a class="" href="directory-interactions.html"><span class="header-section-number">47</span> Directory interactions</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="grouping-data" class="section level1 tabset tabset-fade" number="10">
<h1>
<span class="header-section-number">10</span> Grouping data<a class="anchor" aria-label="anchor" href="#grouping-data"><i class="fas fa-link"></i></a>
</h1>
<p>This page reviews how to group and aggregate data for descriptive analysis. It makes use of tidyverse packages for common and easy-to-use functions.</p>
<!-- ======================================================= -->
<div id="overview-5" class="section level2 tabset tabset-fade tabset-pills" number="10.1">
<h2>
<span class="header-section-number">10.1</span> Overview<a class="anchor" aria-label="anchor" href="#overview-5"><i class="fas fa-link"></i></a>
</h2>
<p>Grouping data is a core component of data management and analysis. Grouped data can be plotted, or summarised by group (whether by time period, place, or a relevant categorical variable). Functions from the <strong>dplyr</strong> package (part of the <strong>tidyverse</strong>) make grouping and subsequent operations quite easy.</p>
<p>This page will how to perform these grouping operations using</p>
<ul>
<li>
<code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> command in <strong>dplyr</strong>
<ul>
<li>Grouping<br>
</li>
<li>Un-group<br>
</li>
<li>Summarise with statistics<br>
</li>
<li>Count and tally<br>
</li>
<li>Arranging grouped data<br>
</li>
<li>Filters applied to grouped data<br>
</li>
<li>Mutate columns within grouped data<br>
</li>
<li>Select columns within grouped data<br>
</li>
</ul>
</li>
<li>Base R <code><a href="https://rdrr.io/r/stats/aggregate.html">aggregate()</a></code> command as an alternative</li>
</ul>
<!-- ======================================================= -->
</div>
<div id="preparation-1" class="section level2 tabset tabset-fade tabset-pills" number="10.2">
<h2>
<span class="header-section-number">10.2</span> Preparation<a class="anchor" aria-label="anchor" href="#preparation-1"><i class="fas fa-link"></i></a>
</h2>
<p><strong>Load packages</strong></p>
<p>Ensure <strong>tidyverse</strong> package is installed and loaded (includes <strong>dplyr</strong>).</p>
<div class="sourceCode" id="cb370"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load</a></span><span class="op">(</span><span class="va">rio</span>,       <span class="co"># to import data</span>
               <span class="va">here</span>,      <span class="co"># to locate files</span>
               <span class="va">tidyverse</span>, <span class="co"># to clean, handle, and plot the data (includes dplyr)</span>
               <span class="va">janitor</span>    <span class="co"># adding total rows and columns</span>
               <span class="op">)</span></code></pre></div>
<p><strong>Load data</strong></p>
<p>For this page we use the cleaned <code>linelist</code> dataset</p>
<div class="sourceCode" id="cb371"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">&lt;-</span> <span class="fu">rio</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rio/man/import.html">import</a></span><span class="op">(</span><span class="fu">here</span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"linelist_cleaned.xlsx"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The first 50 rows of <code>linelist</code>:</p>
<div id="htmlwidget-d25481f3d0e106143f68" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-d25481f3d0e106143f68">{"x":{"filter":"top","filterHTML":"<tr>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"2\" data-max=\"15\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"date\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"1399334400000\" data-max=\"1407283200000\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"date\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"1399507200000\" data-max=\"1407628800000\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"date\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"1.39968e+12\" data-max=\"1407715200000\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"date\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"1400371200000\" data-max=\"1410566400000\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"0\" data-max=\"40\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"disabled\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"0\" data-max=\"40\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"factor\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"width: 100%; display: none;\">\n      <select multiple=\"multiple\" style=\"width: 100%;\" data-options=\"[&quot;0-4&quot;,&quot;5-9&quot;,&quot;10-14&quot;,&quot;15-19&quot;,&quot;20-29&quot;,&quot;30-49&quot;,&quot;50-69&quot;,&quot;70+&quot;]\"><\/select>\n    <\/div>\n  <\/td>\n  <td data-type=\"factor\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"width: 100%; display: none;\">\n      <select multiple=\"multiple\" style=\"width: 100%;\" data-options=\"[&quot;0-4&quot;,&quot;5-9&quot;,&quot;10-14&quot;,&quot;15-19&quot;,&quot;20-24&quot;,&quot;25-29&quot;,&quot;30-34&quot;,&quot;35-39&quot;,&quot;40-44&quot;,&quot;45-49&quot;,&quot;50-54&quot;,&quot;55-59&quot;,&quot;60-64&quot;,&quot;65-69&quot;,&quot;70-74&quot;,&quot;75-79&quot;,&quot;80-84&quot;,&quot;85+&quot;]\"><\/select>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"-13.2688968279886\" data-max=\"-13.2097478342339\" data-scale=\"13\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"8.44852192548534\" data-max=\"8.48604294056283\" data-scale=\"14\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"0\" data-max=\"85\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"22\" data-max=\"203\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"20\" data-max=\"25\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"35.6\" data-max=\"38\" data-scale=\"1\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"0\" data-max=\"228.928199791883\" data-scale=\"14\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"0\" data-max=\"2\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n<\/tr>","data":[["d8a13d","8689b7","11f8ea","dae8c7","acf422","1a4ac9","275cc7","1389ca","057e7a","c97dd9","02d8fd","c36eb4","88eff2","542d07","f50e8a","7f5a01","b799eb","415e3a","bc2adf","188c88","959170","dd9d20","6d788e","67be4e","a9feb3","f81362","4c1040","2cb9a5","342b03","92ddf7","69446e","583717","0a9827","3789ee","d9f1b9","14224c","c5d4ec","53f4ae","c8c030","6b70f0","675f57","1a9636","059cc0","9cd6d0","4ade58","571076","7d99f2","6dca26","579c1d","267434"],[4,4,2,3,6,6,5,4,7,9,9,8,5,5,10,7,5,6,6,10,8,7,11,8,6,8,7,11,7,9,12,10,9,10,8,8,7,7,9,7,11,14,10,12,11,13,12,14,11,15],["2014-05-06",null,null,"2014-05-23","2014-05-25",null,"2014-05-24",null,"2014-06-04",null,"2014-06-14","2014-06-15",null,"2014-06-10",null,"2014-06-23","2014-06-27",null,"2014-07-03","2014-07-03","2014-06-29",null,"2014-07-12","2014-07-15","2014-06-29",null,"2014-07-17",null,null,"2014-07-19",null,null,null,"2014-07-26","2014-07-18","2014-07-17","2014-07-21","2014-07-30",null,null,"2014-07-30",null,null,"2014-08-01","2014-08-02","2014-08-06","2014-07-24",null,"2014-08-03","2014-08-01"],["2014-05-08","2014-05-13","2014-05-16",null,"2014-05-27","2014-05-27","2014-05-27","2014-06-05","2014-06-14",null,"2014-06-20","2014-06-20","2014-06-20",null,"2014-06-22","2014-06-25","2014-07-03","2014-07-06","2014-07-09","2014-07-09","2014-07-13",null,"2014-07-16","2014-07-17","2014-07-19","2014-07-20",null,"2014-07-22","2014-07-22","2014-07-23","2014-07-23","2014-07-27","2014-07-28","2014-08-01","2014-08-02","2014-08-02","2014-08-02",null,"2014-08-02","2014-08-03","2014-08-04","2014-08-05","2014-08-05","2014-08-06","2014-08-07","2014-08-07","2014-08-08","2014-08-09","2014-08-09","2014-08-10"],["2014-05-10","2014-05-14","2014-05-18","2014-05-27","2014-05-28","2014-05-29","2014-05-28","2014-06-07","2014-06-15","2014-06-19","2014-06-20","2014-06-21","2014-06-21","2014-06-23","2014-06-23","2014-06-27","2014-07-05","2014-07-08","2014-07-09","2014-07-11","2014-07-13","2014-07-17","2014-07-17","2014-07-19","2014-07-21","2014-07-21","2014-07-20","2014-07-22","2014-07-24","2014-07-25","2014-07-25","2014-07-28","2014-07-30","2014-08-02","2014-08-03","2014-08-02","2014-08-02","2014-08-04","2014-08-02","2014-08-04","2014-08-04","2014-08-06","2014-08-05","2014-08-07","2014-08-09","2014-08-08","2014-08-10","2014-08-10","2014-08-10","2014-08-11"],[null,"2014-05-18","2014-05-30","2014-05-30","2014-06-27","2014-06-07","2014-06-07","2014-06-09",null,"2014-07-11","2014-07-01","2014-06-24","2014-06-24","2014-06-28","2014-07-01","2014-07-06","2014-07-12","2014-07-26",null,"2014-07-16","2014-07-19","2014-07-21",null,"2014-08-14","2014-08-03","2014-08-07",null,"2014-08-28","2014-08-24",null,"2014-07-27","2014-08-06","2014-07-20","2014-09-13","2014-08-24","2014-08-19",null,null,"2014-08-11",null,"2014-08-18","2014-08-13","2014-08-24","2014-08-13","2014-09-08","2014-08-24","2014-08-27","2014-08-14","2014-08-16","2014-08-21"],[null,"Recover","Recover","Death","Recover","Death","Death","Death","Recover","Recover","Death","Death","Death","Death",null,"Death","Recover","Recover",null,null,"Death","Recover","Recover","Recover","Death","Death","Recover","Recover",null,"Recover","Death","Death",null,null,null,null,null,"Death",null,"Death","Recover","Death","Recover","Recover","Recover","Recover","Recover","Death","Death","Recover"],["f","f","m","f","m","m","f","f","f","m","m","m","f","m","f","f","m","f","m","f","f","m","m","f","f","m",null,"f","f","m","f","m","m","f","m","m","m","f","f","m","f","m","m","m","m","f","f","f","f","m"],[3,7,21,4,4,30,13,2,4,22,20,16,5,16,0,13,14,13,19,3,2,23,13,7,14,35,null,10,9,32,31,27,12,12,21,24,33,9,8,40,28,3,18,17,14,17,1,20,29,1],["years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years","years"],[3,7,21,4,4,30,13,2,4,22,20,16,5,16,0,13,14,13,19,3,2,23,13,7,14,35,null,10,9,32,31,27,12,12,21,24,33,9,8,40,28,3,18,17,14,17,1,20,29,1],["0-4","5-9","20-29","0-4","0-4","30-49","10-14","0-4","0-4","20-29","20-29","15-19","5-9","15-19","0-4","10-14","10-14","10-14","15-19","0-4","0-4","20-29","10-14","5-9","10-14","30-49",null,"10-14","5-9","30-49","30-49","20-29","10-14","10-14","20-29","20-29","30-49","5-9","5-9","30-49","20-29","0-4","15-19","15-19","10-14","15-19","0-4","20-29","20-29","0-4"],["0-4","5-9","20-24","0-4","0-4","30-34","10-14","0-4","0-4","20-24","20-24","15-19","5-9","15-19","0-4","10-14","10-14","10-14","15-19","0-4","0-4","20-24","10-14","5-9","10-14","35-39",null,"10-14","5-9","30-34","30-34","25-29","10-14","10-14","20-24","20-24","30-34","5-9","5-9","40-44","25-29","0-4","15-19","15-19","10-14","15-19","0-4","20-24","25-29","0-4"],["St. Mark's Maternity Hospital (SMMH)","Missing","St. Mark's Maternity Hospital (SMMH)","Port Hospital","Central Hospital","Port Hospital","Central Hospital","Missing","Missing","Port Hospital","Port Hospital","Port Hospital","St. Mark's Maternity Hospital (SMMH)","Port Hospital","Port Hospital","Missing","Missing","Other","Missing","Central Hospital","Central Hospital","Missing","Missing","Other","St. Mark's Maternity Hospital (SMMH)","Other","Other","Port Hospital","Other","Missing","Military Hospital","Other","Port Hospital","St. Mark's Maternity Hospital (SMMH)","Military Hospital","Missing","St. Mark's Maternity Hospital (SMMH)","Military Hospital","Missing","Missing","Port Hospital","Port Hospital","Missing","Port Hospital","Other","Military Hospital","Missing","Port Hospital","Other","Other"],[-13.2354719577436,-13.2152339775486,-13.212910703914,-13.2197320896877,-13.2688968279886,-13.2604684939726,-13.2369393782451,-13.2572163655863,-13.2180326141688,-13.2243437095992,-13.2124311244169,-13.2619688324425,-13.2195578885261,-13.2628934511971,-13.2336087079551,-13.2339681355349,-13.2162953240809,-13.2531041397858,-13.2266742923612,-13.2123762615712,-13.2452992638476,-13.2192936985287,-13.2199077448676,-13.2343062806506,-13.2155668757305,-13.2163318124535,-13.2359603887432,-13.2097478342339,-13.2209544818868,-13.2221173306518,-13.2306657089937,-13.2587625731645,-13.2243932729634,-13.2137356012883,-13.2207093341173,-13.2564700881739,-13.2258087869163,-13.2223378683793,-13.2106014146595,-13.2486407324245,-13.2188046852032,-13.2461808628649,-13.2123634125052,-13.2111276946553,-13.2591405401705,-13.2191390852229,-13.2301859855135,-13.2481679868778,-13.2163350530737,-13.217405738006],[8.4627682729487,8.45171855856465,8.46481700596819,8.44852192548534,8.47845973648553,8.45421479878989,8.46916113869591,8.47292327643506,8.47155535608018,8.47145134147474,8.45344099552146,8.48475855218221,8.46210832824745,8.46479592922727,8.47804840685363,8.46957530395867,8.46395131808455,8.46110718866038,8.48408263734462,8.4594706833607,8.48334624336805,8.4859413033916,8.4693933891765,8.47121232619015,8.4627846208641,8.4630635336683,8.47704542384682,8.47714159984428,8.48436046542975,8.46420961833913,8.47808779372907,8.45574180260702,8.48325844901445,8.4732571907655,8.47571579396786,8.48329801272452,8.47317350834362,8.46136226718465,8.47793008081529,8.48480340615605,8.47782489076703,8.48313501406402,8.48309382244331,8.46975025432843,8.45279482787079,8.44933234263738,8.47770243597866,8.46556229437808,8.48604294056283,8.47529881925793],["20b688",null,null,"be7f8a","1511c5",null,"e02f66",null,"cbbe78",null,"e61cb9","057e7a",null,"4977bd",null,"a75c7f","ea3740",null,"b799eb","9d07e6","beb26e",null,"36e2e7","7baf73","af1c1f",null,"916d0a",null,null,"d15be5",null,null,null,"3b096b","7baf73","d0eead","805a4f","4af3d1",null,null,"7ee655",null,null,"0e3c82","3789ee","9cd6d0","ce4ec6",null,"b56849","db7b54"],["other",null,null,"other","other",null,"other",null,"funeral",null,"other","other",null,"funeral",null,"other","other",null,"other","funeral","funeral",null,"other","other","other",null,"other",null,null,"other",null,null,null,"other","other","other","funeral","other",null,null,"other",null,null,"funeral","other","other","other",null,"other","other"],[33,41,61,42,51,74,55,19,39,69,61,59,37,67,0,49,79,48,70,28,22,67,70,24,53,85,25,51,48,74,71,69,66,49,65,73,67,53,42,74,74,41,64,64,67,58,21,55,63,20],[55,104,171,60,89,173,115,52,67,135,133,149,78,141,22,113,153,142,137,50,31,160,168,84,144,174,42,115,120,201,203,182,140,134,132,166,176,112,100,173,124,62,126,113,164,98,40,130,163,53],[22,22,23,24,24,22,22,23,23,22,22,21,23,23,25,23,21,22,21,20,21,21,21,24,22,21,23,22,23,23,22,21,23,23,22,22,24,22,21,22,22,21,21,23,22,21,22,23,23,22],["no","no","no",null,"no","no","no","no","no","no","no","no","no","no","no","no","no",null,null,"no","no",null,"no","no","no","no","no","no","no","no","no","no","no",null,null,"no","no","no","no",null,"no","no","no","no","no","no","no",null,"no","no"],["no","no","no",null,"no","no","yes","yes","no","no","no","yes","yes","no","no","no","no",null,null,"no","no",null,"yes","no","no","yes","no","no","no","no","no","no","no",null,null,"yes","yes","no","no",null,"yes","yes","no","no","no","yes","yes",null,"no","no"],["no","yes","yes",null,"yes","yes","yes","yes","yes","yes","yes","yes","yes","yes","yes","no","yes",null,null,"yes","yes",null,"yes","no","yes","yes","yes","yes","yes","yes","yes","yes","no",null,null,"yes","yes","yes","yes",null,"yes","no","no","no","yes","yes","yes",null,"yes","yes"],["no","no","no",null,"no","no","no","no","no","no","no","no","no","no","no","no","no",null,null,"no","no",null,"no","no","no","no","no","no","no","no","no","no","no",null,null,"no","no","no","no",null,"no","no","no","no","no","no","no",null,"no","no"],["no","no","no",null,"yes","no","yes","no","no","yes","yes","no","no","no","yes","yes","no",null,null,"yes","yes",null,"yes","yes","yes","no","yes","no","no","no","yes","no","yes",null,null,"yes","yes","no","yes",null,"no","no","no","no","no","yes","no",null,"no","yes"],[37.2,37,37,37.4,36.9,37.3,37.5,37.3,36.8,36.5,36.9,37,37.3,35.6,37.3,37,37,37.1,36.8,36.8,37.5,36.6,36.6,37.1,37.2,37.1,37.1,36.9,37.4,37.7,36.8,36.7,36.3,37.2,37.4,38,36,37.3,37.8,36.3,36.7,37.3,37.5,36,37,37.3,37.4,36.4,37.1,36.5],["19:08","14:47","11:19","17:22","10:37",null,"09:40","10:28","18:15","10:20","21:28","12:31","11:17",null,null,"11:48","15:03","14:26","10:49","14:32","08:48","17:27","23:27","19:03","16:40",null,"18:58","12:28","14:19",null,"16:54","10:11","06:53","09:48","05:56","07:21","14:48","16:29",null,"15:42",null,"15:07","15:45",null,null,"21:54","09:13","05:17","14:30","09:32"],[109.090909090909,37.9068047337278,20.8611196607503,116.666666666667,64.3858098724908,24.7251829329413,41.5879017013233,70.2662721893491,86.879037647583,37.8600823045267,34.4847080106281,26.5753794874105,60.8152531229454,33.7005180825914,0,38.374187485316,33.7477038745782,23.8048006347947,37.2955405189408,112,228.928199791883,26.171875,24.8015873015873,34.0136054421769,25.5594135802469,28.075042938301,141.72335600907,38.5633270321361,33.3333333333333,18.3163783074676,17.2292460384867,20.8308175341142,33.6734693877551,27.288928491869,37.3048668503214,26.4915082014806,21.6296487603306,42.2512755102041,42,24.7251829329413,48.1269510926119,106.659729448491,40.3124212648022,50.121387735923,24.9107674003569,60.3915035401916,131.25,32.5443786982248,23.7118446309609,71.1997152011392],[2,1,2,null,1,2,1,2,1,null,0,1,1,null,1,2,2,2,0,2,0,null,1,2,2,1,null,0,2,2,2,1,2,1,1,0,0,null,0,1,0,1,0,1,2,1,2,1,1,1]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>case_id<\/th>\n      <th>generation<\/th>\n      <th>date_infection<\/th>\n      <th>date_onset<\/th>\n      <th>date_hospitalisation<\/th>\n      <th>date_outcome<\/th>\n      <th>outcome<\/th>\n      <th>gender<\/th>\n      <th>age<\/th>\n      <th>age_unit<\/th>\n      <th>age_years<\/th>\n      <th>age_cat<\/th>\n      <th>age_cat5<\/th>\n      <th>hospital<\/th>\n      <th>lon<\/th>\n      <th>lat<\/th>\n      <th>infector<\/th>\n      <th>source<\/th>\n      <th>wt_kg<\/th>\n      <th>ht_cm<\/th>\n      <th>ct_blood<\/th>\n      <th>fever<\/th>\n      <th>chills<\/th>\n      <th>cough<\/th>\n      <th>aches<\/th>\n      <th>vomit<\/th>\n      <th>temp<\/th>\n      <th>time_admission<\/th>\n      <th>bmi<\/th>\n      <th>days_onset_hosp<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[1,8,10,14,15,18,19,20,26,28,29]}],"order":[],"autoWidth":false,"orderClasses":false,"orderCellsTop":true,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><!-- ======================================================= -->
</div>
<div id="grouping" class="section level2 tabset tabset-fade tabset-pills" number="10.3">
<h2>
<span class="header-section-number">10.3</span> Grouping<a class="anchor" aria-label="anchor" href="#grouping"><i class="fas fa-link"></i></a>
</h2>
<p>The function <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> from <strong>dplyr</strong> groups the rows by the unique values in the specified columns. Each unique value contitutes a group (or unique combination of values, if multiple grouping columns are specified). Subsequent changes to the dataset or calculations can then be performed within the context of each unique group.</p>
<p>For example, the command below takes the linelist and groups the rows by unique values in column <code>outcome</code>, saving the output as a new dataframe <code>ll_by_outcome</code>. The column name is placed inside the parentheses of the function <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code>.</p>
<div class="sourceCode" id="cb372"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ll_by_outcome</span> <span class="op">&lt;-</span> <span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span></code></pre></div>
<p><strong>Note that there is no perceptible change to the dataset</strong> after <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code>, <em>until</em> another <strong>dplyr</strong> verb such as <code><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code> or <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code> is applied on the “grouped” dataframe.</p>
<p>You can however “see” the groupings by printing the dataframe. When you print a grouped dataframe, you will see it has been transformed into a <code>tibble</code> class object (LINK) which, when printed, displays which grouping columns have been applied and how many groups there are - written just above the header row.</p>
<div class="sourceCode" id="cb373"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># print to see which groups are active</span>
<span class="va">ll_by_outcome</span></code></pre></div>
<pre><code>## # A tibble: 5,888 x 30
## # Groups:   outcome [3]
##    case_id generation date_infection date_onset date_hospitalis~ date_outcome outcome gender   age age_unit age_years age_cat age_cat5 hospital   lon   lat
##    &lt;chr&gt;        &lt;dbl&gt; &lt;date&gt;         &lt;date&gt;     &lt;date&gt;           &lt;date&gt;       &lt;chr&gt;   &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt; &lt;fct&gt;   &lt;fct&gt;    &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt;
##  1 d8a13d           4 2014-05-06     2014-05-08 2014-05-10       NA           &lt;NA&gt;    f          3 years            3 0-4     0-4      St. Mar~ -13.2  8.46
##  2 8689b7           4 NA             2014-05-13 2014-05-14       2014-05-18   Recover f          7 years            7 5-9     5-9      Missing  -13.2  8.45
##  3 11f8ea           2 NA             2014-05-16 2014-05-18       2014-05-30   Recover m         21 years           21 20-29   20-24    St. Mar~ -13.2  8.46
##  4 dae8c7           3 2014-05-23     NA         2014-05-27       2014-05-30   Death   f          4 years            4 0-4     0-4      Port Ho~ -13.2  8.45
##  5 acf422           6 2014-05-25     2014-05-27 2014-05-28       2014-06-27   Recover m          4 years            4 0-4     0-4      Central~ -13.3  8.48
##  6 1a4ac9           6 NA             2014-05-27 2014-05-29       2014-06-07   Death   m         30 years           30 30-49   30-34    Port Ho~ -13.3  8.45
##  7 275cc7           5 2014-05-24     2014-05-27 2014-05-28       2014-06-07   Death   f         13 years           13 10-14   10-14    Central~ -13.2  8.47
##  8 1389ca           4 NA             2014-06-05 2014-06-07       2014-06-09   Death   f          2 years            2 0-4     0-4      Missing  -13.3  8.47
##  9 057e7a           7 2014-06-04     2014-06-14 2014-06-15       NA           Recover f          4 years            4 0-4     0-4      Missing  -13.2  8.47
## 10 c97dd9           9 NA             NA         2014-06-19       2014-07-11   Recover m         22 years           22 20-29   20-24    Port Ho~ -13.2  8.47
## # ... with 5,878 more rows, and 14 more variables: infector &lt;chr&gt;, source &lt;chr&gt;, wt_kg &lt;dbl&gt;, ht_cm &lt;dbl&gt;, ct_blood &lt;dbl&gt;, fever &lt;chr&gt;, chills &lt;chr&gt;,
## #   cough &lt;chr&gt;, aches &lt;chr&gt;, vomit &lt;chr&gt;, temp &lt;dbl&gt;, time_admission &lt;chr&gt;, bmi &lt;dbl&gt;, days_onset_hosp &lt;dbl&gt;</code></pre>
<div id="unique-groups" class="section level3" number="10.3.1">
<h3>
<span class="header-section-number">10.3.1</span> Unique groups<a class="anchor" aria-label="anchor" href="#unique-groups"><i class="fas fa-link"></i></a>
</h3>
<p><strong>The groups created reflect each unique combination of values in the grouping columns.</strong> To see the groups and the number of rows in each group, pass the grouped data to <code><a href="https://dplyr.tidyverse.org/reference/count.html">tally()</a></code>.</p>
<p>See below that there are <strong>three</strong> unique values in the grouping column <code>outcome</code>: “Death”, “Recover”, and <code>NA</code>. See that there were 2582 deaths, 1983 recoveries, and 1323 with no outcome recorded.</p>
<div class="sourceCode" id="cb375"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">tally</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 2
##   outcome     n
## * &lt;chr&gt;   &lt;int&gt;
## 1 Death    2582
## 2 Recover  1983
## 3 &lt;NA&gt;     1323</code></pre>
<p>You can group by more than one column. Below, the dataframe is grouped by <code>outcome</code> and <code>gender</code>, and then tallied. Note how each unique combination of <code>outcome</code> and <code>gender</code> is registered as its own group - including missing values for either column.</p>
<div class="sourceCode" id="cb377"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">outcome</span>, <span class="va">gender</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">tally</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 9 x 3
## # Groups:   outcome [3]
##   outcome gender     n
##   &lt;chr&gt;   &lt;chr&gt;  &lt;int&gt;
## 1 Death   f       1236
## 2 Death   m       1211
## 3 Death   &lt;NA&gt;     135
## 4 Recover f        944
## 5 Recover m        947
## 6 Recover &lt;NA&gt;      92
## 7 &lt;NA&gt;    f        631
## 8 &lt;NA&gt;    m        633
## 9 &lt;NA&gt;    &lt;NA&gt;      59</code></pre>
</div>
<div id="new-columns-1" class="section level3" number="10.3.2">
<h3>
<span class="header-section-number">10.3.2</span> New columns<a class="anchor" aria-label="anchor" href="#new-columns-1"><i class="fas fa-link"></i></a>
</h3>
<p>You can also create a new grouping column <em>within</em> the group_by() statement. This is equivalent to calling <code><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code> before the <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code>. For a quick tabulation this style can be handy, but for more clarity in your code consider creating this column in it’s own <code><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code> step and then piping to <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code>.</p>
<div class="sourceCode" id="cb379"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># group dat based on a binary column created *within* the group_by() command</span>
<span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span>
    age_class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">age</span> <span class="op">&gt;=</span> <span class="fl">18</span>, <span class="st">"adult"</span>, <span class="st">"child"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">tally</a></span><span class="op">(</span>sort <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 2
##   age_class     n
##   &lt;chr&gt;     &lt;int&gt;
## 1 child      3598
## 2 adult      2202
## 3 &lt;NA&gt;         88</code></pre>
</div>
<div id="replaceadd-grouping-columns" class="section level3" number="10.3.3">
<h3>
<span class="header-section-number">10.3.3</span> Replace/add grouping columns<a class="anchor" aria-label="anchor" href="#replaceadd-grouping-columns"><i class="fas fa-link"></i></a>
</h3>
<p>By default if you run <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> on data that are already grouped, the old groups will be removed and the new one(s) will apply. If you want to add new groups to the existing ones, add the argument <code>.add=TRUE</code>.</p>
<div class="sourceCode" id="cb381"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Grouped by outcome</span>
<span class="va">by_outcome</span> <span class="op">&lt;-</span> <span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span>

<span class="co"># Add grouping by gender in addition</span>
<span class="va">by_outcome_gender</span> <span class="op">&lt;-</span> <span class="va">by_outcome</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">gender</span>, .add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="un-group" class="section level2" number="10.4">
<h2>
<span class="header-section-number">10.4</span> Un-group<a class="anchor" aria-label="anchor" href="#un-group"><i class="fas fa-link"></i></a>
</h2>
<p>Data that have been grouped will remain grouped until specifically ungrouped via <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup()</a></code>. If you forget to ungroup, it can lead to incorrect calculations! Below is an example of removing all grouping columns:</p>
<div class="sourceCode" id="cb382"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">outcome</span>, <span class="va">gender</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">tally</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>You can also remove grouping by only specific columns, by placing the column name inside.</p>
<div class="sourceCode" id="cb383"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">outcome</span>, <span class="va">gender</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">tally</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="va">gender</span><span class="op">)</span></code></pre></div>
<p><span style="color: black;"><strong><em>NOTE:</em></strong> The verb <code><a href="https://dplyr.tidyverse.org/reference/count.html">count()</a></code> automatically ungroups the data after counting.</span></p>
</div>
<div id="summarise" class="section level2" number="10.5">
<h2>
<span class="header-section-number">10.5</span> Summarise<a class="anchor" aria-label="anchor" href="#summarise"><i class="fas fa-link"></i></a>
</h2>
<p>By applying the <strong>dplyr</strong> verb <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code> to grouped data, you can produce summary tables containing descriptive statistics <em>for each group</em>.</p>
<p>Within the summarise statement, provide the name(s) of the new summary column(s), an equals sign, and then a statistical function to apply to the data, as shown below. Within a statistical function, list the column to be operated on and any relevant arguements. For example, do not forget <code>na.rm=TRUE</code> to remove missing values from calculations!</p>
<p>Below is an example of <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code> <em>applied without grouped data</em>. The statistics returned are produced from the entire dataset.</p>
<div class="sourceCode" id="cb384"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>
    mean_age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">age_years</span>, na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span>,
    max_age  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">age_years</span>, na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span>,
    min_age  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">age_years</span>, na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##            mean_age max_age min_age
## 1 16.06242816091954      77       0</code></pre>
<p>In contrast, below is the same <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code> statement applied to grouped data. The statistics are calculated for each <code>outcome</code> group.</p>
<div class="sourceCode" id="cb386"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>
    mean_age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">age_years</span>, na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span>,
    max_age  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">age_years</span>, na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span>,
    min_age  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">age_years</span>, na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 4
##   outcome mean_age max_age min_age
## * &lt;chr&gt;      &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
## 1 Death       16.5      74       0
## 2 Recover     15.7      77       0
## 3 &lt;NA&gt;        15.7      75       0</code></pre>
<p><span style="color: darkgreen;"><strong><em>TIP:</em></strong> Summarise works with both UK and US spelling - <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize()</a></code> call the same function.</span></p>
<div id="summarise-across-multiple-columns" class="section level3" number="10.5.1">
<h3>
<span class="header-section-number">10.5.1</span> Summarise <code>across()</code> multiple columns<a class="anchor" aria-label="anchor" href="#summarise-across-multiple-columns"><i class="fas fa-link"></i></a>
</h3>
<p>You can use summarise across multiple columns using <code><a href="https://dplyr.tidyverse.org/reference/across.html">across()</a></code>. Provide a vector of column names, or use the same semantic helper functions used in <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code> (look below) to specify columns by or by criteria.</p>
<p>Below, <code><a href="https://rdrr.io/r/base/mean.html">mean()</a></code> is applied to ungrouped data (global calculation). The columns are specified, a function is specified (no parentheses), and finally, any additional arguments for the function (e.g. <code>na.rm=TRUE</code>).</p>
<div class="sourceCode" id="cb388"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span>.cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">age_years</span>, <span class="va">temp</span><span class="op">)</span>,
                   .fns <span class="op">=</span> <span class="va">mean</span>,
                   na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##           age_years              temp
## 1 16.06242816091954 38.56498263888889</code></pre>
<p>Below, the same summarise across call is applied on grouped data:</p>
<div class="sourceCode" id="cb390"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span>.cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">age_years</span>, <span class="va">temp</span><span class="op">)</span>, .fns <span class="op">=</span> <span class="va">mean</span>, na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 3
##   outcome age_years  temp
## * &lt;chr&gt;       &lt;dbl&gt; &lt;dbl&gt;
## 1 Death        16.5  38.5
## 2 Recover      15.7  38.6
## 3 &lt;NA&gt;         15.7  38.6</code></pre>
<p>Here are those <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code> helper functions that you can place <em>within</em> <code><a href="https://dplyr.tidyverse.org/reference/across.html">across()</a></code>:</p>
<p>There are helpers available to assist you in specifying columns:</p>
<ul>
<li>
<code><a href="https://tidyselect.r-lib.org/reference/everything.html">everything()</a></code> - all other columns not mentioned<br>
</li>
<li>
<code><a href="https://tidyselect.r-lib.org/reference/everything.html">last_col()</a></code> - the last column<br>
</li>
<li>
<code>where()</code> - applies a function to all columns and selects those which are TRUE<br>
</li>
<li>
<code><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with()</a></code> - matches to a specified prefix. Example: <code><a href="https://dplyr.tidyverse.org/reference/select.html">select(starts_with("date"))</a></code>
</li>
<li>
<code><a href="https://tidyselect.r-lib.org/reference/starts_with.html">ends_with()</a></code> - matches to a specified suffix. Example: <code><a href="https://dplyr.tidyverse.org/reference/select.html">select(ends_with("_end"))</a></code><br>
</li>
<li>
<code><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains()</a></code> - columns containing a character string. Example: <code><a href="https://dplyr.tidyverse.org/reference/select.html">select(contains("time"))</a></code>
</li>
<li>
<code><a href="https://tidyselect.r-lib.org/reference/starts_with.html">matches()</a></code> - to apply a regular expression (regex). Example: <code><a href="https://dplyr.tidyverse.org/reference/select.html">select(contains("[pt]al"))</a></code><br>
</li>
<li>
<code><a href="https://tidyselect.r-lib.org/reference/starts_with.html">num_range()</a></code> -</li>
<li>
<code><a href="https://tidyselect.r-lib.org/reference/all_of.html">any_of()</a></code> - matches if column is named. Useful if the name might not exist. Example: <code><a href="https://dplyr.tidyverse.org/reference/select.html">select(any_of(date_onset, date_death, cardiac_arrest))</a></code>
</li>
</ul>
<p>For example, to return the mean of every numeric column:</p>
<div class="sourceCode" id="cb392"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span>, .fns <span class="op">=</span> <span class="va">mean</span>, na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 12
##   outcome generation   age age_years   lon   lat wt_kg ht_cm ct_blood  temp   bmi days_onset_hosp
## * &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;           &lt;dbl&gt;
## 1 Death         16.7  16.6      16.5 -13.2  8.47  54.0  127.     21.3  38.5  46.8            1.83
## 2 Recover       16.4  15.8      15.7 -13.2  8.47  52.4  124.     21.1  38.6  49.0            2.33
## 3 &lt;NA&gt;          16.5  15.7      15.7 -13.2  8.47  52.3  125.     21.2  38.6  47.6            2.08</code></pre>
<p>If you want summary multiple statistics of multiple columns, in an easy-to-read format, consider a two-way table with the <strong>gtsummary</strong> package This package is demonstrated more extensively in the the Statistics page (LINK).</p>
<div class="sourceCode" id="cb394"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ddsjoberg/gtsummary">gtsummary</a></span><span class="op">)</span> 

<span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">outcome</span>, <span class="va">age_years</span>, <span class="va">temp</span>, <span class="va">ht_cm</span><span class="op">)</span> <span class="op">%&gt;%</span>     <span class="co"># select columns (optional)</span>
  <span class="fu">gtsummary</span><span class="fu">::</span><span class="fu"><a href="http://www.danieldsjoberg.com/gtsummary//reference/tbl_summary.html">tbl_summary</a></span><span class="op">(</span> 
    by <span class="op">=</span> <span class="va">outcome</span>,                                 <span class="co"># indicate grouping column (optional)</span>
    statistic <span class="op">=</span> <span class="fu"><a href="http://www.danieldsjoberg.com/gtsummary//reference/select_helpers.html">all_continuous</a></span><span class="op">(</span><span class="op">)</span> <span class="op">~</span> <span class="st">"{mean} ({sd})"</span><span class="op">)</span> <span class="co"># return mean and std deviation for each group</span></code></pre></div>
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#dqheatuzcr .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#dqheatuzcr .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#dqheatuzcr .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#dqheatuzcr .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 4px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#dqheatuzcr .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#dqheatuzcr .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#dqheatuzcr .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#dqheatuzcr .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#dqheatuzcr .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#dqheatuzcr .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#dqheatuzcr .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#dqheatuzcr .gt_group_heading {
  padding: 8px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#dqheatuzcr .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#dqheatuzcr .gt_from_md > :first-child {
  margin-top: 0;
}

#dqheatuzcr .gt_from_md > :last-child {
  margin-bottom: 0;
}

#dqheatuzcr .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#dqheatuzcr .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 12px;
}

#dqheatuzcr .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#dqheatuzcr .gt_first_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
}

#dqheatuzcr .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#dqheatuzcr .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#dqheatuzcr .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#dqheatuzcr .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#dqheatuzcr .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#dqheatuzcr .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding: 4px;
}

#dqheatuzcr .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#dqheatuzcr .gt_sourcenote {
  font-size: 90%;
  padding: 4px;
}

#dqheatuzcr .gt_left {
  text-align: left;
}

#dqheatuzcr .gt_center {
  text-align: center;
}

#dqheatuzcr .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#dqheatuzcr .gt_font_normal {
  font-weight: normal;
}

#dqheatuzcr .gt_font_bold {
  font-weight: bold;
}

#dqheatuzcr .gt_font_italic {
  font-style: italic;
}

#dqheatuzcr .gt_super {
  font-size: 65%;
}

#dqheatuzcr .gt_footnote_marks {
  font-style: italic;
  font-size: 65%;
}
</style>
<div id="dqheatuzcr" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;"><div class="inline-table"><table class="gt_table">
<thead class="gt_col_headings"><tr>
<th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1"><strong>Characteristic</strong></th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">
<strong>Death</strong>, N = 2,582<sup class="gt_footnote_marks">1</sup>
</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_center" rowspan="1" colspan="1">
<strong>Recover</strong>, N = 1,983<sup class="gt_footnote_marks">1</sup>
</th>
    </tr></thead>
<tbody class="gt_table_body">
<tr>
<td class="gt_row gt_left">age_years</td>
      <td class="gt_row gt_center">17 (13)</td>
      <td class="gt_row gt_center">16 (13)</td>
    </tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">Unknown</td>
      <td class="gt_row gt_center">48</td>
      <td class="gt_row gt_center">23</td>
    </tr>
<tr>
<td class="gt_row gt_left">temp</td>
      <td class="gt_row gt_center">38.55 (1.00)</td>
      <td class="gt_row gt_center">38.59 (0.96)</td>
    </tr>
<tr>
<td class="gt_row gt_left" style="text-align: left; text-indent: 10px;">Unknown</td>
      <td class="gt_row gt_center">51</td>
      <td class="gt_row gt_center">53</td>
    </tr>
<tr>
<td class="gt_row gt_left">ht_cm</td>
      <td class="gt_row gt_center">127 (49)</td>
      <td class="gt_row gt_center">124 (50)</td>
    </tr>
</tbody>
<tfoot><tr class="gt_footnotes">
<td colspan="3">
        <p class="gt_footnote">
          <sup class="gt_footnote_marks"><em>1</em>
          </sup>
           
          Mean (SD)
          <br></p>
      </td>
    </tr></tfoot>
</table></div></div>
</div>
</div>
<div id="count-and-tally" class="section level2" number="10.6">
<h2>
<span class="header-section-number">10.6</span> Count and tally<a class="anchor" aria-label="anchor" href="#count-and-tally"><i class="fas fa-link"></i></a>
</h2>
<p><code><a href="https://dplyr.tidyverse.org/reference/count.html">count()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/count.html">tally()</a></code> provide similar functionality but are different.</p>
<p><code><a href="https://dplyr.tidyverse.org/reference/count.html">tally()</a></code> is shorthand for <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code>, and <em>does not</em> automatically group data. Thus, to achieve grouped tallys it must follow a <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> command. You can add <code>sort = TRUE</code> to see the largest groups first.</p>
<div class="sourceCode" id="cb395"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="va">tally</span></code></pre></div>
<pre><code>##      n
## 1 5888</code></pre>
<div class="sourceCode" id="cb397"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">tally</a></span><span class="op">(</span>sort <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 2
##   outcome     n
##   &lt;chr&gt;   &lt;int&gt;
## 1 Death    2582
## 2 Recover  1983
## 3 &lt;NA&gt;     1323</code></pre>
<p>In contrast, <code><a href="https://dplyr.tidyverse.org/reference/count.html">count()</a></code> does the following:</p>
<ul>
<li>applies <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> on the specified column(s)<br>
</li>
<li>applies <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code> and returned column <code>n</code> with the number of observations per group<br>
</li>
<li>applies <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup()</a></code>
</li>
</ul>
<div class="sourceCode" id="cb399"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span></code></pre></div>
<pre><code>##   outcome    n
## 1   Death 2582
## 2 Recover 1983
## 3    &lt;NA&gt; 1323</code></pre>
<p>Just like with <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> you can create a new column within the <code><a href="https://dplyr.tidyverse.org/reference/count.html">count()</a></code> command:</p>
<div class="sourceCode" id="cb401"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span>age_class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">age</span> <span class="op">&gt;=</span> <span class="fl">18</span>, <span class="st">"adult"</span>, <span class="st">"child"</span><span class="op">)</span>, sort <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<pre><code>##   age_class    n
## 1     child 3598
## 2     adult 2202
## 3      &lt;NA&gt;   88</code></pre>
<p>Read more about the distinction between <code><a href="https://dplyr.tidyverse.org/reference/count.html">tally()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/count.html">count()</a></code> <a href="https://dplyr.tidyverse.org/reference/tally.html">here</a></p>
<p>Both of these verbs can be called multiple times, with the functionality “rolling up”. For example, to summarise the number of genders present for each outcome, run the following. Note, the name of the final column is changed from default “n” for clarity.</p>
<div class="sourceCode" id="cb403"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="co"># produce counts by outcome-gender groups</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">outcome</span>, <span class="va">gender</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># produce counts of gender within each outcome group</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">outcome</span>, name <span class="op">=</span> <span class="st">"number of genders per outcome"</span> <span class="op">)</span> </code></pre></div>
<pre><code>##   outcome number of genders per outcome
## 1   Death                             3
## 2 Recover                             3
## 3    &lt;NA&gt;                             3</code></pre>
<div id="add-totals" class="section level3" number="10.6.1">
<h3>
<span class="header-section-number">10.6.1</span> Add totals<a class="anchor" aria-label="anchor" href="#add-totals"><i class="fas fa-link"></i></a>
</h3>
<p>If you want to add total rows or column after using <code><a href="https://dplyr.tidyverse.org/reference/count.html">tally()</a></code> or <code><a href="https://dplyr.tidyverse.org/reference/count.html">count()</a></code>, consider using the <strong>janitor</strong> package, which offers functions like <code>adorn_totals()</code> and <code>adorn_percentages()</code>. There are many useful functions (search Help for details), here are a few of them:</p>
<ul>
<li>Use <code>adorn_totals()</code> to get totals - specify the argument <code>where =</code> either “row” or “col” or <code><a href="https://rdrr.io/r/base/c.html">c("row", "col")</a></code>.<br>
</li>
<li>Use <code>adorn_percentages()</code> to convert counts to proportions - specify the argument <code>denominator =</code> either “row”, “col”, or “all”.<br>
</li>
<li>Use <code>adorn_pct_formatting()</code> to convert proportions to percentages (can specify number of <code>digits =</code>, whether to add “%” with <code>affix_sign =</code>, and specify specific column names to operate on)<br>
</li>
<li>Use <code>adorn_ns()</code> to add back the underlying counts (“N”s) to a table whose proportions were calculated by <code>adorn_percentages()</code> - to display them together. Indicate <code>position =</code> of the Ns as either “rear” or “front” of the proportions.</li>
</ul>
<p>To add totals:</p>
<div class="sourceCode" id="cb405"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">adorn_totals</span><span class="op">(</span>where <span class="op">=</span> <span class="st">"col"</span><span class="op">)</span></code></pre></div>
<pre><code>##  outcome    n Total
##    Death 2582  2582
##  Recover 1983  1983
##     &lt;NA&gt; 1323  1323</code></pre>
<p>To convert the numbers to proportions:</p>
<div class="sourceCode" id="cb407"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">adorn_totals</span><span class="op">(</span>where <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span> <span class="op">%&gt;%</span>              <span class="co"># add total row</span>
  <span class="fu">adorn_percentages</span><span class="op">(</span>denominator <span class="op">=</span> <span class="st">"col"</span><span class="op">)</span> <span class="op">%&gt;%</span>   <span class="co"># convert to proportions</span>
  <span class="fu">adorn_rounding</span><span class="op">(</span>digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>                   <span class="co"># round the proportions</span></code></pre></div>
<pre><code>##  outcome    n
##    Death 0.44
##  Recover 0.34
##     &lt;NA&gt; 0.22
##    Total 1.00</code></pre>
<p><strong>janitor</strong> functions can be use together, as below:</p>
<div class="sourceCode" id="cb409"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">outcome</span><span class="op">)</span> <span class="op">%&gt;%</span>              <span class="co"># produce the counts by unique outcome</span>
  <span class="fu">adorn_totals</span><span class="op">(</span>where <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># add total row</span>
  <span class="fu">adorn_percentages</span><span class="op">(</span><span class="st">"col"</span><span class="op">)</span> <span class="op">%&gt;%</span>    <span class="co"># add proportion by column</span>
  <span class="fu">adorn_pct_formatting</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>      <span class="co"># proportion converted to percent</span>
  <span class="fu">adorn_ns</span><span class="op">(</span>position <span class="op">=</span> <span class="st">"front"</span><span class="op">)</span>    <span class="co"># Add the underlying N, in front of the percentage</span></code></pre></div>
<pre><code>##  outcome             n
##    Death 2582  (43.9%)
##  Recover 1983  (33.7%)
##     &lt;NA&gt; 1323  (22.5%)
##    Total 5888 (100.0%)</code></pre>
</div>
</div>
<div id="arranging-grouped-data" class="section level2" number="10.7">
<h2>
<span class="header-section-number">10.7</span> Arranging grouped data<a class="anchor" aria-label="anchor" href="#arranging-grouped-data"><i class="fas fa-link"></i></a>
</h2>
<p>TO DO</p>
<p>Using the <strong>dplyr</strong> verb <code><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange()</a></code> to order the rows in a dataframe behaves the same when the data are grouped, *unless you set the argument <code>.by_group =TRUE</code>. In this case the rows are ordered first by the grouping columns and then by any other columns you specify.</p>
</div>
<div id="filter-on-grouped-data" class="section level2" number="10.8">
<h2>
<span class="header-section-number">10.8</span> Filter on grouped data<a class="anchor" aria-label="anchor" href="#filter-on-grouped-data"><i class="fas fa-link"></i></a>
</h2>
<div id="filter" class="section level3" number="10.8.1">
<h3>
<span class="header-section-number">10.8.1</span> <code>filter()</code><a class="anchor" aria-label="anchor" href="#filter"><i class="fas fa-link"></i></a>
</h3>
<p>When applied in conjunction with functions that evaluate the dataframe (like <code><a href="https://rdrr.io/r/base/Extremes.html">max()</a></code>, <code><a href="https://rdrr.io/r/base/Extremes.html">min()</a></code>, <code><a href="https://rdrr.io/r/base/mean.html">mean()</a></code>), these functions will now be applied to the groups. For example, if you want to filter and keep rows where patients are above the median age, this will now apply per group.</p>
<p>TO DO MORE</p>
</div>
<div id="slice-rows-per-group" class="section level3" number="10.8.2">
<h3>
<span class="header-section-number">10.8.2</span> Slice rows per group<a class="anchor" aria-label="anchor" href="#slice-rows-per-group"><i class="fas fa-link"></i></a>
</h3>
<p>The <strong>dplyr</strong> function <code><a href="https://dplyr.tidyverse.org/reference/slice.html">slice()</a></code>, which <a href="https://dplyr.tidyverse.org/reference/slice.html">subsets rows based on their position</a> in the data, can also be applied per group. Remember to account for sorting the data within each group to get the desired “slice”.</p>
<p>For example, to retrieve only the latest 5 admissions from each hospital:</p>
<ol style="list-style-type: decimal">
<li>Group the linelist by column <code>hospital</code><br>
</li>
<li>Arrange the records from latest to earliest <code>date_hospitalisation</code> <em>within each hospital group</em><br>
</li>
<li>Slice to retrieve the first 5 rows</li>
</ol>
<div class="sourceCode" id="cb411"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">hospital</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">hospital</span>, <span class="va">date_hospitalisation</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">hospital</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">case_id</span>, <span class="va">hospital</span>, <span class="va">date_hospitalisation</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 30 x 3
## # Groups:   hospital [6]
##    case_id hospital          date_hospitalisation
##    &lt;chr&gt;   &lt;chr&gt;             &lt;date&gt;              
##  1 20b688  Central Hospital  2014-05-06          
##  2 d58402  Central Hospital  2014-05-10          
##  3 b8f2fd  Central Hospital  2014-05-13          
##  4 acf422  Central Hospital  2014-05-28          
##  5 275cc7  Central Hospital  2014-05-28          
##  6 d1fafd  Military Hospital 2014-04-17          
##  7 974bc1  Military Hospital 2014-05-13          
##  8 6a9004  Military Hospital 2014-05-13          
##  9 09e386  Military Hospital 2014-05-14          
## 10 865581  Military Hospital 2014-05-15          
## # ... with 20 more rows</code></pre>
<p><code><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_head()</a></code> - selects n rows from the top<br><code><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_tail()</a></code> - selects n rows from the end<br><code><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_sample()</a></code> - randomly selects n rows<br><code><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_min()</a></code> - selects n rows with highest values in <code>order_by =</code> column, use <code>with_ties = TRUE</code> to keep ties<br><code><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max()</a></code> - selects n rows with lowest values in <code>order_by =</code> column, use <code>with_ties = TRUE</code> to keep ties</p>
</div>
<div id="filter-on-group-size" class="section level3" number="10.8.3">
<h3>
<span class="header-section-number">10.8.3</span> Filter on group size<a class="anchor" aria-label="anchor" href="#filter-on-group-size"><i class="fas fa-link"></i></a>
</h3>
<p>The function <code><a href="https://dplyr.tidyverse.org/reference/count.html">add_count()</a></code> adds a column <code>n</code> to the original data giving the number of rows in that row’s group.</p>
<p>Shown below for simplicity is a selection of the <code>linelist</code> data - <code><a href="https://dplyr.tidyverse.org/reference/count.html">add_count()</a></code> is applied to <code>hospital</code>, so the values in column <code>n</code> reflect the number of rows in that row’s hospital group. Note how values are repeated. In the example below, the column name <code>n</code> could be changed to <code>name =</code></p>
<div class="sourceCode" id="cb413"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">case_id</span>, <span class="va">hospital</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">add_count</a></span><span class="op">(</span><span class="va">hospital</span><span class="op">)</span> <span class="op">%&gt;%</span>          <span class="co"># add "number of rows admitted to same hospital as this row" </span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>                          <span class="co"># show just the first 10 rows, for demo purposes</span></code></pre></div>
<pre><code>##    case_id                             hospital    n
## 1   d8a13d St. Mark's Maternity Hospital (SMMH)  422
## 2   8689b7                              Missing 1469
## 3   11f8ea St. Mark's Maternity Hospital (SMMH)  422
## 4   dae8c7                        Port Hospital 1762
## 5   acf422                     Central Hospital  454
## 6   1a4ac9                        Port Hospital 1762
## 7   275cc7                     Central Hospital  454
## 8   1389ca                              Missing 1469
## 9   057e7a                              Missing 1469
## 10  c97dd9                        Port Hospital 1762</code></pre>
<p>It then becomes easy to filter for case rows who were hospitalized at a “small” hospital, say, a hospital that admitted fewer than 500 patients:</p>
<div class="sourceCode" id="cb415"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">case_id</span>, <span class="va">hospital</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">add_count</a></span><span class="op">(</span><span class="va">hospital</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">n</span> <span class="op">&lt;</span> <span class="fl">500</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="mutate-on-grouped-data" class="section level2" number="10.9">
<h2>
<span class="header-section-number">10.9</span> Mutate on grouped data<a class="anchor" aria-label="anchor" href="#mutate-on-grouped-data"><i class="fas fa-link"></i></a>
</h2>
<p>To retain all columns and rows (not summarize) and <em>add a new variable for average group statistics</em>, use <code><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code> instead of <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code>.</p>
<p>This is useful if you want group statistics in the original dataset with all other column present - e.g. for calculations comparing one row to the group.</p>
<p>For example, this code below calculates the difference between a row’s delay-to-admission and the median delay for their hospital. The steps are:</p>
<ol style="list-style-type: decimal">
<li>Group the data by hospital<br>
</li>
<li>Use the column <code>days_onset_hosp</code> (delay to hospitalisation) to create a new column containing the mean delay at the hospital of <em>that row</em><br>
</li>
<li>Calculate the difference between the two columns</li>
</ol>
<div class="sourceCode" id="cb416"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">linelist</span> <span class="op">%&gt;%</span> 
  <span class="co"># group data by hospital (no change to linelist yet)</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">hospital</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  
  <span class="co"># new columns</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>
    <span class="co"># mean days to admission per hospital (rounded to 1 decimal)</span>
    group_delay_admit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">days_onset_hosp</span>, na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span>,
    
    <span class="co"># difference between row's delay and mean delay at their hospital (rounded to 1 decimal)</span>
    diff_to_group     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">days_onset_hosp</span> <span class="op">-</span> <span class="va">group_delay_admit</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># select certain rows only - for demonstration/viewing purposes</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">case_id</span>, <span class="va">hospital</span>, <span class="va">days_onset_hosp</span>, <span class="va">group_delay_admit</span>, <span class="va">diff_to_group</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 5,888 x 5
## # Groups:   hospital [6]
##    case_id hospital                             days_onset_hosp group_delay_admit diff_to_group
##    &lt;chr&gt;   &lt;chr&gt;                                          &lt;dbl&gt;             &lt;dbl&gt;         &lt;dbl&gt;
##  1 d8a13d  St. Mark's Maternity Hospital (SMMH)               2               2.1          -0.1
##  2 8689b7  Missing                                            1               2.1          -1.1
##  3 11f8ea  St. Mark's Maternity Hospital (SMMH)               2               2.1          -0.1
##  4 dae8c7  Port Hospital                                     NA               2.1          NA  
##  5 acf422  Central Hospital                                   1               1.9          -0.9
##  6 1a4ac9  Port Hospital                                      2               2.1          -0.1
##  7 275cc7  Central Hospital                                   1               1.9          -0.9
##  8 1389ca  Missing                                            2               2.1          -0.1
##  9 057e7a  Missing                                            1               2.1          -1.1
## 10 c97dd9  Port Hospital                                     NA               2.1          NA  
## # ... with 5,878 more rows</code></pre>
</div>
<div id="select-on-grouped-data" class="section level2" number="10.10">
<h2>
<span class="header-section-number">10.10</span> Select on grouped data<a class="anchor" aria-label="anchor" href="#select-on-grouped-data"><i class="fas fa-link"></i></a>
</h2>
<p>The verb <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code> works on grouped data, but the grouping columns are always included (even if not mentioned in <code><a href="https://dplyr.tidyverse.org/reference/select.html">select()</a></code>).<br>
If you do not want these grouping columns, use <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup()</a></code> first.</p>
<!-- ======================================================= -->
</div>
<div id="base-r" class="section level2 tabset tabset-fade tabset-pills" number="10.11">
<h2>
<span class="header-section-number">10.11</span> Base R<a class="anchor" aria-label="anchor" href="#base-r"><i class="fas fa-link"></i></a>
</h2>
<p>Here we briefly demonstrate grouping data with the <strong>base</strong> R function <code><a href="https://rdrr.io/r/stats/aggregate.html">aggregate()</a></code></p>
<p>TO DO</p>
<!-- ======================================================= -->
</div>
<div id="resources-2" class="section level2 tabset tabset-fade tabset-pills" number="10.12">
<h2>
<span class="header-section-number">10.12</span> Resources<a class="anchor" aria-label="anchor" href="#resources-2"><i class="fas fa-link"></i></a>
</h2>
<p>Here are some useful resources for more information:
<em><a href="https://github.com/rstudio/cheatsheets/blob/master/data-transformation.pdf" class="uri">https://github.com/rstudio/cheatsheets/blob/master/data-transformation.pdf</a>
</em><a href="https://datacarpentry.org/R-genomics/04-dplyr.html" class="uri">https://datacarpentry.org/R-genomics/04-dplyr.html</a>
<em><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="uri">https://dplyr.tidyverse.org/reference/group_by.html</a>
<a href="https://dplyr.tidyverse.org/articles/grouping.html" class="uri">https://dplyr.tidyverse.org/articles/grouping.html</a><br></em><a href="https://itsalocke.com/files/DataManipulationinR.pdf" class="uri">https://itsalocke.com/files/DataManipulationinR.pdf</a></p>
<p>You can perform any summary function on grouped data; see the Cheat Sheet here for more info:
<a href="https://github.com/rstudio/cheatsheets/blob/master/data-transformation.pdf" class="uri">https://github.com/rstudio/cheatsheets/blob/master/data-transformation.pdf</a></p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="missing-data.html"><span class="header-section-number">9</span> Missing data</a></div>
<div class="next"><a href="joining-data.html"><span class="header-section-number">11</span> Joining data</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#grouping-data"><span class="header-section-number">10</span> Grouping data</a></li>
<li><a class="nav-link" href="#overview-5"><span class="header-section-number">10.1</span> Overview</a></li>
<li><a class="nav-link" href="#preparation-1"><span class="header-section-number">10.2</span> Preparation</a></li>
<li>
<a class="nav-link" href="#grouping"><span class="header-section-number">10.3</span> Grouping</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#unique-groups"><span class="header-section-number">10.3.1</span> Unique groups</a></li>
<li><a class="nav-link" href="#new-columns-1"><span class="header-section-number">10.3.2</span> New columns</a></li>
<li><a class="nav-link" href="#replaceadd-grouping-columns"><span class="header-section-number">10.3.3</span> Replace/add grouping columns</a></li>
</ul>
</li>
<li><a class="nav-link" href="#un-group"><span class="header-section-number">10.4</span> Un-group</a></li>
<li>
<a class="nav-link" href="#summarise"><span class="header-section-number">10.5</span> Summarise</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#summarise-across-multiple-columns"><span class="header-section-number">10.5.1</span> Summarise across() multiple columns</a></li></ul>
</li>
<li>
<a class="nav-link" href="#count-and-tally"><span class="header-section-number">10.6</span> Count and tally</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#add-totals"><span class="header-section-number">10.6.1</span> Add totals</a></li></ul>
</li>
<li><a class="nav-link" href="#arranging-grouped-data"><span class="header-section-number">10.7</span> Arranging grouped data</a></li>
<li>
<a class="nav-link" href="#filter-on-grouped-data"><span class="header-section-number">10.8</span> Filter on grouped data</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#filter"><span class="header-section-number">10.8.1</span> filter()</a></li>
<li><a class="nav-link" href="#slice-rows-per-group"><span class="header-section-number">10.8.2</span> Slice rows per group</a></li>
<li><a class="nav-link" href="#filter-on-group-size"><span class="header-section-number">10.8.3</span> Filter on group size</a></li>
</ul>
</li>
<li><a class="nav-link" href="#mutate-on-grouped-data"><span class="header-section-number">10.9</span> Mutate on grouped data</a></li>
<li><a class="nav-link" href="#select-on-grouped-data"><span class="header-section-number">10.10</span> Select on grouped data</a></li>
<li><a class="nav-link" href="#base-r"><span class="header-section-number">10.11</span> Base R</a></li>
<li><a class="nav-link" href="#resources-2"><span class="header-section-number">10.12</span> Resources</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>R Handbook for Epidemiologists</strong>" was written by the handbook team. It was last built on 2021-02-20.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
