<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>38 Phylogenetic trees | The Epidemiologist R Handbook</title>
<meta name="author" content="the handbook team">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.2"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.7/header-attrs.js"></script><script src="libs/jquery-3.5.1/jquery-3.5.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.5.3/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.5.3/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.2.3.9000/tabs.js"></script><script src="libs/bs3compat-0.2.3.9000/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.15/datatables.js"></script><link href="libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="libs/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet">
<script src="libs/nouislider-7.0.10/jquery.nouislider.min.js"></script><link href="libs/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet">
<script src="libs/selectize-0.12.0/selectize.min.js"></script><link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script><script src="libs/kePrint-0.0.1/kePrint.js"></script><link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<link href="libs/tabwid-1.0.0/tabwid.css" rel="stylesheet">
<link href="libs/leaflet-1.3.1/leaflet.css" rel="stylesheet">
<script src="libs/leaflet-1.3.1/leaflet.js"></script><link href="libs/leafletfix-1.0.0/leafletfix.css" rel="stylesheet">
<script src="libs/proj4-2.6.2/proj4.min.js"></script><script src="libs/Proj4Leaflet-1.0.1/proj4leaflet.js"></script><link href="libs/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet">
<script src="libs/leaflet-binding-2.0.4.1/leaflet.js"></script><script src="libs/leaflet-providers-1.9.0/leaflet-providers_1.9.0.js"></script><script src="libs/leaflet-providers-plugin-2.0.4.1/leaflet-providers-plugin.js"></script><script src="libs/viz-1.8.2/viz.js"></script><link href="libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="libs/grViz-binding-1.0.6.1/grViz.js"></script><script src="libs/d3-4.5.0/d3.min.js"></script><script src="libs/sankey-1/sankey.js"></script><script src="libs/sankeyNetwork-binding-0.4/sankeyNetwork.js"></script><script src="libs/plotly-binding-4.9.3/plotly.js"></script><script src="libs/typedarray-0.1/typedarray.min.js"></script><link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script><link href="libs/vis-4.20.1/vis.css" rel="stylesheet">
<script src="libs/vis-4.20.1/vis.min.js"></script><script src="libs/visNetwork-binding-2.0.9/visNetwork.js"></script><link href="libs/font-awesome-4.7.0/css/font-awesome.min.css" rel="stylesheet">
<script src="libs/plotly-basic-1.57.1/plotly-basic-1.57.1.min.js"></script><script src="libs/plotly-cartesian-1.57.1/plotly-cartesian-1.57.1.min.js"></script><script src="https://cdn.jsdelivr.net/autocomplete.js/0/autocomplete.jquery.min.js"></script><script src="https://cdn.jsdelivr.net/npm/mark.js@8.11.1/dist/mark.min.js"></script><!-- CSS --><link rel="stylesheet" href="style_bs4.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">The Epidemiologist R Handbook</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"></a></li>
<li class="book-part">About this book</li>
<li><a class="" href="editorial-and-technical-notes.html"><span class="header-section-number">1</span> Editorial and technical notes</a></li>
<li><a class="" href="download-book-and-data.html"><span class="header-section-number">2</span> Download book and data</a></li>
<li class="book-part">Basics</li>
<li><a class="" href="r-basics.html"><span class="header-section-number">3</span> R Basics</a></li>
<li><a class="" href="transition-to-r.html"><span class="header-section-number">4</span> Transition to R</a></li>
<li><a class="" href="import-and-export.html"><span class="header-section-number">5</span> Import and export</a></li>
<li><a class="" href="r-projects.html"><span class="header-section-number">6</span> R projects</a></li>
<li><a class="" href="suggested-packages-1.html"><span class="header-section-number">7</span> Suggested packages</a></li>
<li class="book-part">Data Management</li>
<li><a class="" href="cleaning-data-and-core-functions.html"><span class="header-section-number">8</span> Cleaning data and core functions</a></li>
<li><a class="" href="working-with-dates.html"><span class="header-section-number">9</span> Working with dates</a></li>
<li><a class="" href="factors.html"><span class="header-section-number">10</span> Factors</a></li>
<li><a class="" href="pivoting-data.html"><span class="header-section-number">11</span> Pivoting data</a></li>
<li><a class="" href="grouping-data.html"><span class="header-section-number">12</span> Grouping data</a></li>
<li><a class="" href="joining-data.html"><span class="header-section-number">13</span> Joining data</a></li>
<li><a class="" href="characters-and-strings.html"><span class="header-section-number">14</span> Characters and strings</a></li>
<li><a class="" href="de-duplication.html"><span class="header-section-number">15</span> De-duplication</a></li>
<li><a class="" href="iteration-and-loops.html"><span class="header-section-number">16</span> Iteration and loops</a></li>
<li class="book-part">Analysis</li>
<li><a class="" href="descriptive-tables.html"><span class="header-section-number">17</span> Descriptive tables</a></li>
<li><a class="" href="simple-statistical-tests.html"><span class="header-section-number">18</span> Simple statistical tests</a></li>
<li><a class="" href="univariate-and-multivariate-regression.html"><span class="header-section-number">19</span> Univariate and multivariate regression</a></li>
<li><a class="" href="missing-data.html"><span class="header-section-number">20</span> Missing data</a></li>
<li><a class="" href="standardization.html"><span class="header-section-number">21</span> Standardization</a></li>
<li><a class="" href="moving-averages.html"><span class="header-section-number">22</span> Moving averages</a></li>
<li><a class="" href="time-series-and-outbreak-detection.html"><span class="header-section-number">23</span> Time series and outbreak detection</a></li>
<li><a class="" href="epidemic-modeling.html"><span class="header-section-number">24</span> Epidemic modeling</a></li>
<li><a class="" href="survey-analysis.html"><span class="header-section-number">25</span> Survey analysis</a></li>
<li><a class="" href="survival-analysis.html"><span class="header-section-number">26</span> Survival analysis</a></li>
<li><a class="" href="gis-basics.html"><span class="header-section-number">27</span> GIS basics</a></li>
<li class="book-part">Data Visualization</li>
<li><a class="" href="ggplot-tips.html"><span class="header-section-number">28</span> ggplot tips</a></li>
<li><a class="" href="epidemic-curves.html"><span class="header-section-number">29</span> Epidemic curves</a></li>
<li><a class="" href="plot-continuous-data.html"><span class="header-section-number">30</span> Plot continuous data</a></li>
<li><a class="" href="plot-categorical-data.html"><span class="header-section-number">31</span> Plot categorical data</a></li>
<li><a class="" href="html-tables.html"><span class="header-section-number">32</span> HTML tables</a></li>
<li><a class="" href="age-pyramids-and-likert-scales.html"><span class="header-section-number">33</span> Age pyramids and Likert-scales</a></li>
<li><a class="" href="diagrams-and-charts.html"><span class="header-section-number">34</span> Diagrams and charts</a></li>
<li><a class="" href="combinations-analysis.html"><span class="header-section-number">35</span> Combinations analysis</a></li>
<li><a class="" href="heat-tiles.html"><span class="header-section-number">36</span> Heat tiles</a></li>
<li><a class="" href="transmission-chains.html"><span class="header-section-number">37</span> Transmission chains</a></li>
<li><a class="active" href="phylogenetic-trees.html"><span class="header-section-number">38</span> Phylogenetic trees</a></li>
<li><a class="" href="interactive-plots.html"><span class="header-section-number">39</span> Interactive plots</a></li>
<li class="book-part">Advanced</li>
<li><a class="" href="directory-interactions.html"><span class="header-section-number">40</span> Directory interactions</a></li>
<li><a class="" href="r-markdown-1.html"><span class="header-section-number">41</span> R Markdown</a></li>
<li><a class="" href="routine-reports.html"><span class="header-section-number">42</span> Routine reports</a></li>
<li><a class="" href="errors-warnings-1.html"><span class="header-section-number">43</span> Errors &amp; warnings</a></li>
<li><a class="" href="advanced-rstudio.html"><span class="header-section-number">44</span> Advanced RStudio</a></li>
<li><a class="" href="relational-databases.html"><span class="header-section-number">45</span> Relational databases</a></li>
<li><a class="" href="shiny-and-dashboards.html"><span class="header-section-number">46</span> Shiny and dashboards</a></li>
<li><a class="" href="collaboration-with-github.html"><span class="header-section-number">47</span> Collaboration with Github</a></li>
<li><a class="" href="writing-functions.html"><span class="header-section-number">48</span> Writing functions</a></li>
<li><a class="" href="r-on-network-drives.html"><span class="header-section-number">49</span> R on network drives</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="phylogenetic-trees" class="section level1" number="38">
<h1>
<span class="header-section-number">38</span> Phylogenetic trees<a class="anchor" aria-label="anchor" href="#phylogenetic-trees"><i class="fas fa-link"></i></a>
</h1>
<p><strong>Phylogenetic trees</strong> are used to visualize and describe the relatedness and evolution of organisms based on the sequence of their genetic code. They can be constructed from genetic sequences using distance-based methods (such as neighbor-joining method) or character-based methods (such as maximum likelihood and Bayesian Markov Chain Monte Carlo method). Next-generation sequencing (NGS) has become more affordable and is becoming more widely used in public health to describe pathogens causing infectious diseases. Portable devices decrease the turn around time and make data available for the support of outbreak investigation in real-time. NGS data can be used to identify the origin or source of an outbreak strain and its propagation, as well as determine presence of antimicrobial resistance genes. To visualize the genetic relatedness between samples a phylogenetic tree is constructed. In this page we will learn how to use the <strong>ggtree()</strong> package, which allows for combination of phylogenetic trees with additional sample data in form of a dataframe in order to help observe patterns and improve understanding of the outbreak dynamic.</p>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_overview_graph-1.png" width="80%" style="display: block; margin: auto;"></div>
<!-- ======================================================= -->
<div id="preparation-30" class="section level2" number="38.1">
<h2>
<span class="header-section-number">38.1</span> Preparation<a class="anchor" aria-label="anchor" href="#preparation-30"><i class="fas fa-link"></i></a>
</h2>
<p>This code chunk shows the loading of packages required for the analyses. In this handbook we emphasize <code>p_load()</code> from <strong>pacman</strong>, which installs the package if necessary and loads it for use. You can also load packages with <code><a href="https://rdrr.io/r/base/library.html">library()</a></code> from <strong>base</strong> R. See the page on <a href="r-basics.html#r-basics">R basics</a> for more information on R packages.</p>
<div class="sourceCode" id="cb1497"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load/install packages</span>
<span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load</a></span><span class="op">(</span><span class="va">here</span>, <span class="va">ggplot2</span>, <span class="va">dplyr</span>, <span class="va">ape</span>, <span class="va">ggtree</span>, <span class="va">treeio</span>, <span class="va">ggnewscale</span><span class="op">)</span></code></pre></div>
<p>There are several different formats in which a phylogenetic tree can be stored (eg. Newick, NEXUS, Phylip). A common one, which we will also use here in this example is the Newick file format (.nwk), which is the standard for representing trees in computer-readable form. Which means, an entire tree can be expressed in a string format such as “((t2:0.04,t1:0.34):0.89,(t5:0.37,(t4:0.03,t3:0.67):0.9):0.59);” listing all nodes and tips and their relationship (branch length) to each other.</p>
<p>It is important to understand that the phylogenetic tree file in itself does not contain sequencing data, but is merely the result of the distances between the sequences. We therefore cannot extract sequencing data from a tree file.</p>
<p>We use the <strong>ape()</strong> package to import a phylogenetic tree file and store it in a list object of class “phylo”. We inspect our tree object and see it contains 299 tips (or samples) and 236 nodes.</p>
<div class="sourceCode" id="cb1498"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># read in the tree: we use the here package to specify the location of our R project and data files:</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html">read.tree</a></span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"Shigella_tree.nwk"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># inspect the tree file:</span>
<span class="va">tree</span></code></pre></div>
<pre><code>## 
## Phylogenetic tree with 299 tips and 236 internal nodes.
## 
## Tip labels:
##   SRR5006072, SRR4192106, S18BD07865, S18BD00489, S17BD08906, S17BD05939, ...
## Node labels:
##   17, 29, 100, 67, 100, 100, ...
## 
## Rooted; includes branch lengths.</code></pre>
<p>Second we import a table with additional information for each sequenced sample such as gender, country of origin and attributes for antimicrobial resistance:</p>
<div class="sourceCode" id="cb1500"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># We read in a csv file into a dataframe format:</span>
<span class="va">sample_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"sample_data_Shigella_tree.csv"</span>, sep <span class="op">=</span> <span class="st">","</span>, na.strings <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NA"</span><span class="op">)</span>, head <span class="op">=</span> <span class="cn">TRUE</span>, stringsAsFactors<span class="op">=</span><span class="cn">F</span><span class="op">)</span></code></pre></div>
<p>Below are the first 30 rows of these data:</p>
<div id="htmlwidget-95c1394d02ee56f34d13" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-95c1394d02ee56f34d13">{"x":{"filter":"top","filterHTML":"<tr>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"disabled\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"disabled\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"disabled\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"disabled\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"disabled\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"integer\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"0\" data-max=\"64\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"number\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n    <div style=\"display: none; position: absolute; width: 200px;\">\n      <div data-min=\"0\" data-max=\"8\" data-scale=\"2\"><\/div>\n      <span style=\"float: left;\"><\/span>\n      <span style=\"float: right;\"><\/span>\n    <\/div>\n  <\/td>\n<\/tr>","data":[["S17BD05944","S15BD07413","S18BD07247","S19BD07384","S18BD07338","S18BD02657","S19BD06884","S16BD02574","S16BD04069","S17BD08170","S15BD07250","S17BD07935","S17BD00384","S19BD06034","S17BD08948","S17BD04542","S17BD07399","S17BD08233","S19BD07381","S17BD07961","S18BD09176","S19BD07393","S18BD02580","S17BD02526","S17BD06833","S18BD01024","S18BD03411","S18BD09515","S17BD07332","S18BD06598"],["Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei","Sonnei"],["Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium","Belgium"],["Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe","Europe"],[null,"Georgia",null,"Morocco",null,null,null,"Thailand",null,null,null,null,null,null,null,"Iran",null,null,null,null,null,null,null,null,null,null,null,null,null,null],["2017","2015","2018","2019","2018","2018","2019","2016","2016","2017","2015","2017","2017","2019","2017","2017","2017","2017","2019","2017","2018","2019","2018","2017","2017","2018","2018","2018","2017","2018"],["Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes","Yes"],["NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL","NRC BEL"],[null,"Male","Female","Female","Male","Male","Male","Male","Male","Male","Female",null,"Male","Male","Female","Male","Male","Female","Female","Male","Male","Male","Male","Female","Male","Female","Female","Male","Female","Male"],[null,null,null,null,null,"gyrA D87Y","gyrA D87Y","gyrA D87Y","gyrA D87Y",null,null,"gyrA D87Y",null,null,"gyrA D87Y",null,null,null,null,null,null,null,null,null,null,"gyrA D87Y","gyrA S83L",null,null,null],[null,null,null,null,null,"mphA","mphA",null,null,null,null,"mphA",null,null,"mphA",null,null,null,null,null,null,null,null,null,null,"mphA","mphA",null,null,null],[8,0,16,8,8,64,64,0,0,64,0,64,8,8,64,8,8,8,8,8,8,16,8,8,8,64,64,8,8,8],[0.25,0,0.03,0.02,0.02,0.5,0.5,0,0,8,0,0.5,0.02,0.02,0.5,0.25,0.02,0.02,0.02,0.02,0.02,0.12,0.02,0.02,0.02,0.5,0.25,0.03,0.02,0.02]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>Sample_ID<\/th>\n      <th>serotype<\/th>\n      <th>Country<\/th>\n      <th>Continent<\/th>\n      <th>Travel_history<\/th>\n      <th>Year<\/th>\n      <th>Belgium<\/th>\n      <th>Source<\/th>\n      <th>Gender<\/th>\n      <th>gyrA_mutations<\/th>\n      <th>macrolide_resistance_genes<\/th>\n      <th>MIC_AZM<\/th>\n      <th>MIC_CIP<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[11,12]}],"order":[],"autoWidth":false,"orderClasses":false,"orderCellsTop":true,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p>We clean and inspect our data: In order to assign the correct sample data to the phylogenetic tree, the Sample_IDs in the sample_data file need to match the tip.labels in the tree file:</p>
<div class="sourceCode" id="cb1501"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># We clean the data: we select certain columns to be protected from cleaning in order to main tain their formating (eg. for the sample names, as they have to match the names in the phylogenetic tree file)</span>
<span class="co">#sample_data &lt;- linelist::clean_data(sample_data, protect = c(1, 3:5)) </span>

<span class="co"># We check the formatting of the tip labels in the tree file: </span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span><span class="op">)</span> <span class="co"># these are the sample names in the tree - we inspect the first 6 with head()</span></code></pre></div>
<pre><code>## [1] "SRR5006072" "SRR4192106" "S18BD07865" "S18BD00489" "S17BD08906" "S17BD05939"</code></pre>
<div class="sourceCode" id="cb1503"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># We make sure the first column in our dataframe are the Sample_IDs:</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sample_data</span><span class="op">)</span>   </code></pre></div>
<pre><code>##  [1] "Sample_ID"                  "serotype"                  
##  [3] "Country"                    "Continent"                 
##  [5] "Travel_history"             "Year"                      
##  [7] "Belgium"                    "Source"                    
##  [9] "Gender"                     "gyrA_mutations"            
## [11] "macrolide_resistance_genes" "MIC_AZM"                   
## [13] "MIC_CIP"</code></pre>
<div class="sourceCode" id="cb1505"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># We look at the sample_IDs in the dataframe to make sure the formatting is the same than in the tip.labels (eg. letters are all capital, no extra _ between letters and numbers etc.)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span><span class="op">)</span> <span class="co"># we inspect only the first 6 using head()</span></code></pre></div>
<pre><code>## [1] "S17BD05944" "S15BD07413" "S18BD07247" "S19BD07384" "S18BD07338" "S18BD02657"</code></pre>
<p>Upon inspection we can see that the format of sample_ID in the dataframe corresponds to the format of sample names at the tree tips. These do not have to be sorted in the same order to be matched.</p>
<p>We are ready to go!</p>
<!-- ======================================================= -->
</div>
<div id="simple-tree-visualization" class="section level2" number="38.2">
<h2>
<span class="header-section-number">38.2</span> Simple tree visualization<a class="anchor" aria-label="anchor" href="#simple-tree-visualization"><i class="fas fa-link"></i></a>
</h2>
<div id="different-tree-layouts" class="section level3 unnumbered">
<h3>Different tree layouts<a class="anchor" aria-label="anchor" href="#different-tree-layouts"><i class="fas fa-link"></i></a>
</h3>
<p><strong>ggtree()</strong> offers many different layout formats and some may be more suitable for your specific purpose than others:</p>
<div class="sourceCode" id="cb1507"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Examples:</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> <span class="co"># most simple linear tree</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>,  branch.length <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="co"># most simple linear tree with all tips aligned</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>, layout<span class="op">=</span><span class="st">"circular"</span><span class="op">)</span> <span class="co"># most simple circular tree</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>, layout<span class="op">=</span><span class="st">"circular"</span>, branch.length <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="co"># most simple circular tree with all tips aligned</span>

<span class="co"># for other options see online: http://yulab-smu.top/treedata-book/chapter4.html</span></code></pre></div>
<p><img src="_main_files/figure-html/phylogenetic_trees_example_formats-1.png" width="50%"><img src="_main_files/figure-html/phylogenetic_trees_example_formats-2.png" width="50%"><img src="_main_files/figure-html/phylogenetic_trees_example_formats-3.png" width="50%"><img src="_main_files/figure-html/phylogenetic_trees_example_formats-4.png" width="50%"></p>
</div>
<div id="simple-tree-with-addition-of-sample-data" class="section level3 unnumbered">
<h3>Simple tree with addition of sample data<a class="anchor" aria-label="anchor" href="#simple-tree-with-addition-of-sample-data"><i class="fas fa-link"></i></a>
</h3>
<p>The most easy annotation of your tree is the addition of the sample names at the tips, as well as coloring of tip points and if desired branches:</p>
<div class="sourceCode" id="cb1508"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># A: Plot Circular tree:</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>, layout<span class="op">=</span><span class="st">"circular"</span>, branch.length<span class="op">=</span><span class="st">'none'</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span> <span class="co"># the %&lt;+% is used to add your dataframe with sample data to the tree</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/AsIs.html">I</a></span><span class="op">(</span><span class="va">Belgium</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>     <span class="co"># color the branches according to a variable in your dataframe</span>
  <span class="fu">scale_color_manual</span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"Sample Origin"</span>,  <span class="co"># name of your color scheme (will show up in the legend like this)</span>
    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Yes"</span>, <span class="st">"No"</span><span class="op">)</span>, <span class="co"># the different options in your variable</span>
    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NRCSS Belgium"</span>, <span class="st">"Other"</span><span class="op">)</span>, <span class="co"># how you want the different options named in your legend, allows for formatting</span>
    values<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"grey"</span><span class="op">)</span>,            <span class="co"># the color you want to assign to the variable if its "nrc_bel"</span>
    na.value<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span><span class="op">+</span>     <span class="co"># for the NA values we choose the color grey</span>
  <span class="fu">new_scale_color</span><span class="op">(</span><span class="op">)</span><span class="op">+</span>      <span class="co"># allows to add an additional color scheme for another variable</span>
     <span class="fu">geom_tippoint</span><span class="op">(</span>       <span class="co"># color the tip point by continent, you may change shape adding "shape = "</span>
       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">Continent</span><span class="op">)</span>,
       size<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span><span class="op">+</span> 
  <span class="fu">scale_color_brewer</span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"Continent"</span>,   <span class="co"># name of your color scheme (will show up in the legend like this)</span>
    palette<span class="op">=</span><span class="st">"Set1"</span>,       <span class="co"># we choose a premade set of colors coming with the brewer package</span>
    na.value<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span><span class="op">+</span>     <span class="co"># for the NA values we choose the color grey</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>            <span class="co"># add the name of the sample to the tip of its branch (you can add as many text lines as you like with the + , you just need to change the offset value to place them next to each other)</span>
    color<span class="op">=</span><span class="st">'black'</span>,
    offset <span class="op">=</span> <span class="fl">1</span>,
    size <span class="op">=</span> <span class="fl">1</span>,
    geom <span class="op">=</span> <span class="st">"text"</span>,
    align<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">+</span> 
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Phylogenetic tree of Shigella sonnei"</span><span class="op">)</span><span class="op">+</span> <span class="co"># title of your graph</span>
  <span class="fu">theme</span><span class="op">(</span>
    axis.title.x<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, <span class="co"># removes x-axis title</span>
    axis.title.y<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, <span class="co"># removes y-axis title</span>
    legend.title<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>face<span class="op">=</span><span class="st">"bold"</span>, size <span class="op">=</span><span class="fl">12</span><span class="op">)</span>,  <span class="co"># defines font size and format of the legend title</span>
    legend.text<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>face<span class="op">=</span><span class="st">"bold"</span>, size <span class="op">=</span><span class="fl">10</span><span class="op">)</span>,   <span class="co"># defines font size and format of the legend text</span>
    plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span><span class="fl">12</span>, face<span class="op">=</span><span class="st">"bold"</span><span class="op">)</span>,  <span class="co"># defines font size and format of the plot title</span>
    legend.position<span class="op">=</span><span class="st">"bottom"</span>, <span class="co"># defines placement of the legend</span>
    legend.box<span class="op">=</span><span class="st">"vertical"</span>,
    legend.margin<span class="op">=</span><span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>   <span class="co"># defines placement of the legend</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_adding_sampledata-1.png" width="120%" style="display: block; margin: auto;"></div>
<div class="sourceCode" id="cb1509"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Export your tree graph:</span>
<span class="fu">ggsave</span><span class="op">(</span><span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"images"</span>, <span class="st">"example_tree_circular_1.png"</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">12</span>, height <span class="op">=</span> <span class="fl">14</span><span class="op">)</span></code></pre></div>
<!-- ======================================================= -->
</div>
</div>
<div id="manipulation-of-phylogenetic-trees" class="section level2" number="38.3">
<h2>
<span class="header-section-number">38.3</span> Manipulation of phylogenetic trees<a class="anchor" aria-label="anchor" href="#manipulation-of-phylogenetic-trees"><i class="fas fa-link"></i></a>
</h2>
<p>Sometimes you may have a very large phylogenetic tree and you are only interested in one part of the tree. For example if you produced a tree including historical or international samples to get a large overview of where your dataset might fit in in the bigger picture. But then to look closer at your data you want to inspect only that portion of the bigger tree.</p>
<p>Since the phylogenetic tree file is just the output of sequencing data analysis, we can not manipulate the order of the nodes and branches in the file itself. These have already been determined in previous analysis from the raw NGS data. We are able though to zoom into parts, hide parts and seven subset part of the tree.</p>
<div id="zooming-in-on-one-part-of-the-tree" class="section level3" number="38.3.1">
<h3>
<span class="header-section-number">38.3.1</span> Zooming in on one part of the tree:<a class="anchor" aria-label="anchor" href="#zooming-in-on-one-part-of-the-tree"><i class="fas fa-link"></i></a>
</h3>
<p>If you don’t want to “cut” your tree, but only inspect part of it more closely you can zoom in to view a specific part:</p>
<div class="sourceCode" id="cb1510"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># First we plot the whole tree:</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>,<span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span><span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span> <span class="co"># labels the tips of all branche with the sample name in the tree file</span>
  <span class="fu">geom_text2</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset<span class="op">=</span><span class="op">!</span><span class="va">isTip</span>, label<span class="op">=</span><span class="va">node</span><span class="op">)</span>, size <span class="op">=</span><span class="fl">5</span>, color <span class="op">=</span> <span class="st">"darkred"</span>, hjust<span class="op">=</span><span class="fl">1</span>, vjust <span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="co"># labels all the nodes in the tree</span>
<span class="va">p</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_zoom_in-1.png" width="70%" style="display: block; margin: auto;"></div>
<p>We want to zoom into the branch which is sticking out, after node number 452 to get a closer look:</p>
<div class="sourceCode" id="cb1511"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">viewClade</span><span class="op">(</span><span class="va">p</span> , node<span class="op">=</span><span class="fl">452</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_zoom_in_452-1.png" width="70%" style="display: block; margin: auto;"></div>
</div>
<div id="collapsing-one-part-of-the-tree" class="section level3" number="38.3.2">
<h3>
<span class="header-section-number">38.3.2</span> Collapsing one part of the tree:<a class="anchor" aria-label="anchor" href="#collapsing-one-part-of-the-tree"><i class="fas fa-link"></i></a>
</h3>
<p>The other way around we may want to ignore this branch which is sticking out and can do so by collapsing it at the node (indicated here by the blue square):</p>
<div class="sourceCode" id="cb1512"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#First we collapse at node 452</span>
<span class="va">p_collapsed</span> <span class="op">&lt;-</span> <span class="fu">collapse</span><span class="op">(</span><span class="va">p</span>, node<span class="op">=</span><span class="fl">452</span><span class="op">)</span>

<span class="co">#To not forget that we collapsed this node we assign a symbol to it:</span>
<span class="va">p_collapsed</span> <span class="op">+</span> <span class="fu">geom_point2</span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset<span class="op">=</span><span class="op">(</span><span class="va">node</span> <span class="op">==</span> <span class="fl">452</span><span class="op">)</span><span class="op">)</span>, size<span class="op">=</span><span class="fl">5</span>, shape<span class="op">=</span><span class="fl">23</span>, fill<span class="op">=</span><span class="st">"steelblue"</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_collapse_452-1.png" width="70%" style="display: block; margin: auto;"></div>
</div>
<div id="subsetting-a-tree" class="section level3 unnumbered">
<h3>Subsetting a tree<a class="anchor" aria-label="anchor" href="#subsetting-a-tree"><i class="fas fa-link"></i></a>
</h3>
<p>If we want to make a more permanent change and create a new tree to work with we can subset part of it and even save it as new newick tree file.</p>
<div class="sourceCode" id="cb1513"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># To do so you can add the node and tip labels to your tree to see which part you want to subset:</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">tree</span>, branch.length<span class="op">=</span><span class="st">'none'</span>, layout<span class="op">=</span><span class="st">'circular'</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span>       <span class="co"># labels the tips of all branche with the sample name in the tree file</span>
  <span class="fu">geom_text2</span><span class="op">(</span>                  <span class="co"># labels all the nodes in the tree</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset <span class="op">=</span> <span class="op">!</span><span class="va">isTip</span>, label<span class="op">=</span><span class="va">node</span><span class="op">)</span>,
    size <span class="op">=</span> <span class="fl">3</span>,
    color <span class="op">=</span> <span class="st">"darkred"</span><span class="op">)</span><span class="op">+</span>  
 <span class="fu">theme</span><span class="op">(</span>
   legend.position <span class="op">=</span> <span class="st">"none"</span>, <span class="co"># removes the legend all together</span>
   axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
   axis.title.y<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
   plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span><span class="fl">12</span>, face<span class="op">=</span><span class="st">"bold"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># A: Subset tree based on node:</span>
<span class="va">sub_tree1</span> <span class="op">&lt;-</span> <span class="fu">tree_subset</span><span class="op">(</span>
  <span class="va">tree</span>,
  node <span class="op">=</span> <span class="fl">528</span><span class="op">)</span> <span class="co"># we subset the tree at node 528</span>

<span class="co"># lets have a look at the subset tree:</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree1</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Subset tree 1"</span><span class="op">)</span>

<span class="co"># B: Subset the same part of the tree based on a samplem in this case S17BD07692:</span>
<span class="va">sub_tree2</span> <span class="op">&lt;-</span> <span class="fu">tree_subset</span><span class="op">(</span>
  <span class="va">tree</span>,
  <span class="st">"S17BD07692"</span>,
  levels_back <span class="op">=</span> <span class="fl">9</span><span class="op">)</span> <span class="co"># levels back defines how many nodes backwards from the sample tip you want to go</span>

<span class="co"># lets have a look at the subset tree:</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Subset tree 2"</span><span class="op">)</span></code></pre></div>
<p><img src="_main_files/figure-html/phylogenetic_trees_subsetting-1.png" width="70%" style="display: block; margin: auto;"><img src="_main_files/figure-html/phylogenetic_trees_subsetting-2.png" width="70%" style="display: block; margin: auto;"><img src="_main_files/figure-html/phylogenetic_trees_subsetting-3.png" width="70%" style="display: block; margin: auto;"></p>
<p>You can also save your new tree as a Newick file:</p>
<div class="sourceCode" id="cb1514"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.tree.html">write.tree</a></span><span class="op">(</span><span class="va">sub_tree2</span>, file<span class="op">=</span><span class="st">'data/Shigelle_subtree_2.nwk'</span><span class="op">)</span></code></pre></div>
</div>
<div id="rotating-nodes-in-a-tree" class="section level3 unnumbered">
<h3>Rotating nodes in a tree<a class="anchor" aria-label="anchor" href="#rotating-nodes-in-a-tree"><i class="fas fa-link"></i></a>
</h3>
<p>As mentioned before we cannot change the order of tips or nodes in the tree, as this is based on their genetic relatedness and is not subject to visual manipulation. But we can rote branches around nodes if that eases our visualization.</p>
<p>First we plot our new subsetted tree with node labels to choose the node we want to manipulate:</p>
<div class="sourceCode" id="cb1515"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_text2</span><span class="op">(</span>                       <span class="co"># label all the nodes in the tree</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>subset<span class="op">=</span><span class="op">!</span><span class="va">isTip</span>, label<span class="op">=</span><span class="va">node</span><span class="op">)</span>,
    size <span class="op">=</span> <span class="fl">5</span>,
    color <span class="op">=</span> <span class="st">"darkred"</span>,
    hjust <span class="op">=</span> <span class="fl">1</span>,
    vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> 
<span class="va">p</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_rotating_1-1.png" width="70%" style="display: block; margin: auto;"></div>
<p>We choose to manipulate node number 39: we do so by applying <strong>ggtree::rotate()</strong> or <strong>ggtree::flip()</strong> indirectly to node 36 so node 39 moves to the bottom and nodes 37 and 38 move to the top:</p>
<div class="sourceCode" id="cb1516"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># </span>
<span class="co"># p1 &lt;- p + geom_hilight(39, "steelblue", extend =0.0015)+ # highlights the node 39 in blue</span>
<span class="co">#    geom_hilight(37, "yellow", extend =0.0015)  + # highlights the node 37 in yellow</span>
<span class="co">#   ggtitle("Original tree")</span>
<span class="co"># </span>
<span class="co"># # we want to rotate node 36 so node 39 is on the bottom and nodes 37 and 38 move to the top:</span>
<span class="co"># # </span>
<span class="co"># rotate(p1, 39) %&gt;% rotate(37)+</span>
<span class="co">#   ggtitle("Rotated Node 36")</span>
<span class="co"># </span>
<span class="co"># # #or we can use the flip command to achieve the same thing:</span>
<span class="co"># flip(p1, 39, 37)</span></code></pre></div>
</div>
<div id="example-subtree-with-sample-data-annotation" class="section level3" number="38.3.3">
<h3>
<span class="header-section-number">38.3.3</span> Example subtree with sample data annotation:<a class="anchor" aria-label="anchor" href="#example-subtree-with-sample-data-annotation"><i class="fas fa-link"></i></a>
</h3>
<p>Lets say we are investigating the cluster of cases with clonal expansion which occured in 2017 and 2018 at node 39 in our sub-tree. We add the year of strain isolation as well as travel history and color by country to see origin of other closely related strains:</p>
<div class="sourceCode" id="cb1517"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Add sample data:</span>
<span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span> 
  <span class="fu">geom_tiplab</span><span class="op">(</span>                      <span class="co"># labels the tips of all branches with the sample name in the tree file</span>
    size <span class="op">=</span><span class="fl">2.5</span>,
    offset <span class="op">=</span> <span class="fl">0.001</span>, 
    align <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">theme_tree2</span><span class="op">(</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">xlab</span><span class="op">(</span><span class="st">"genetic distance"</span><span class="op">)</span><span class="op">+</span> <span class="co"># add a label to the x-azis</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.015</span><span class="op">)</span><span class="op">+</span>           <span class="co"># set the x-axis limits of our tree</span>
  <span class="fu">geom_tippoint</span><span class="op">(</span>            <span class="co"># color the tip point by continent</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color<span class="op">=</span><span class="va">Country</span><span class="op">)</span>,
    size<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span><span class="op">+</span> 
  <span class="fu">scale_color_brewer</span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"Country"</span>, 
    palette<span class="op">=</span><span class="st">"Set1"</span>, 
    na.value<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>              <span class="co"># add isolation year</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">Year</span><span class="op">)</span>,
    color<span class="op">=</span><span class="st">'blue'</span>,
    offset <span class="op">=</span> <span class="fl">0.0045</span>,
    size <span class="op">=</span> <span class="fl">3</span>,
    linetype <span class="op">=</span> <span class="st">"blank"</span>,
    geom <span class="op">=</span> <span class="st">"text"</span>, 
    align<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>              <span class="co"># add travel history</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">Travel_history</span><span class="op">)</span>,
    color<span class="op">=</span><span class="st">'red'</span>,
    offset <span class="op">=</span> <span class="fl">0.006</span>,
    size <span class="op">=</span> <span class="fl">3</span>,
    linetype <span class="op">=</span> <span class="st">"blank"</span>,
    geom <span class="op">=</span> <span class="st">"text"</span>,
    align<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">+</span> 
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Phylogenetic tree of Belgian S. sonnei strains with travel history"</span><span class="op">)</span><span class="op">+</span> <span class="co"># add plot title</span>
  <span class="fu">theme</span><span class="op">(</span>
    axis.title.x<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
    axis.title.y<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
    legend.title<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>face<span class="op">=</span><span class="st">"bold"</span>, size <span class="op">=</span><span class="fl">12</span><span class="op">)</span>,
    legend.text<span class="op">=</span><span class="fu">element_text</span><span class="op">(</span>face<span class="op">=</span><span class="st">"bold"</span>, size <span class="op">=</span><span class="fl">10</span><span class="op">)</span>,
    plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span><span class="fl">12</span>, face<span class="op">=</span><span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_inspect_subset_example-1.png" width="80%" style="display: block; margin: auto;"></div>
<p>Our observation points towards an import of strains from Asia, which then circulated in Belgium over the years and seem to have caused our latest outbreak.</p>
<!-- ======================================================= -->
</div>
</div>
<div id="more-complex-trees-adding-heatmaps-of-sample-data" class="section level2" number="38.4">
<h2>
<span class="header-section-number">38.4</span> More complex trees: adding heatmaps of sample data<a class="anchor" aria-label="anchor" href="#more-complex-trees-adding-heatmaps-of-sample-data"><i class="fas fa-link"></i></a>
</h2>
<p>We can add more complex information, such as categorical presence of antimicrobial resistance genes and numeric values for actually measured resistance to antimicrobials in form of a heatmap using the <strong>ggtree::gheatmap()</strong> function.</p>
<p>First we need to plot our tree (this can be either linear or circular): We will use the sub_stree from part 3.)</p>
<div class="sourceCode" id="cb1518"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># A: Circular tree:</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span>, branch.length<span class="op">=</span><span class="st">'none'</span>, layout<span class="op">=</span><span class="st">'circular'</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">theme</span><span class="op">(</span>
    legend.position <span class="op">=</span> <span class="st">"none"</span>,
    axis.title.x<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
    axis.title.y<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
    plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span><span class="fl">12</span>, face<span class="op">=</span><span class="st">"bold"</span>,hjust <span class="op">=</span> <span class="fl">0.5</span>, vjust <span class="op">=</span> <span class="op">-</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span>

<span class="va">p</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap-1.png" width="60%" style="display: block; margin: auto;"></div>
<p>Second we prepare our data. To visualize different variables with new color schemes, we subset our dataframe to the desired variable.</p>
<p>For example we want to look at gender and mutations that could confer resistance to ciprofloxacin:</p>
<div class="sourceCode" id="cb1519"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create your gender dataframe:</span>
<span class="va">gender</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"gender"</span> <span class="op">=</span> <span class="va">sample_data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Gender"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>

<span class="co"># Its important to add the Sample_ID as rownames otherwise it cannot match the data to the tree tip.labels:</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gender</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span>

<span class="co"># Create your ciprofloxacin dataframe based on mutations in the gyrA gene:</span>
<span class="va">cipR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"cipR"</span> <span class="op">=</span> <span class="va">sample_data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gyrA_mutations"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">cipR</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span>

<span class="co"># Create your ciprofloxacin dataframe based on the measured minimum inhibitory concentration (MIC) from the laboratory:</span>
<span class="va">MIC_Cip</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"mic_cip"</span> <span class="op">=</span> <span class="va">sample_data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MIC_CIP"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">MIC_Cip</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_data</span><span class="op">$</span><span class="va">Sample_ID</span></code></pre></div>
<p>We create a first plot adding a binary heatmap for gender to the phylogenetic tree:</p>
<div class="sourceCode" id="cb1520"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># First we add gender:</span>
<span class="va">h1</span> <span class="op">&lt;-</span>  <span class="fu">gheatmap</span><span class="op">(</span>
  <span class="va">p</span>,
  <span class="va">gender</span>,
  offset <span class="op">=</span> <span class="fl">12</span>,        <span class="co"># offset shifts the heatmap to the right</span>
  width<span class="op">=</span><span class="fl">0.10</span>,         <span class="co"># width defines the width of the heatmap column</span>
  color<span class="op">=</span><span class="cn">NULL</span>,         <span class="co"># color defines the border of the heatmap columns</span>
  colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span>  <span class="co"># hides column names for the heatmap</span>
  <span class="fu">scale_fill_manual</span><span class="op">(</span>  <span class="co"># define the coloring scheme and legend for gender</span>
    name <span class="op">=</span> <span class="st">"Gender"</span>, 
    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#00d1b1"</span>, <span class="st">"purple"</span><span class="op">)</span>,
    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span>,
    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>
    legend.position<span class="op">=</span><span class="st">"bottom"</span>,
    legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,
    legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span><span class="fl">10</span><span class="op">)</span>,
    legend.box<span class="op">=</span><span class="st">"vertical"</span>, legend.margin<span class="op">=</span><span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">h1</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_gender-1.png" width="70%" style="display: block; margin: auto;"></div>
<p>Then we add information on ciprofloxacin resistance genes:</p>
<div class="sourceCode" id="cb1521"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># First we assigng a new color scheme to our existing plot, this enables us to define and change the colors for our second variable</span>
<span class="va">h2</span> <span class="op">&lt;-</span> <span class="va">h1</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span> 

<span class="co"># then we combine these into a new plot:</span>
<span class="va">h3</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span>
  <span class="va">h2</span>,
  <span class="va">cipR</span>,
  offset <span class="op">=</span> <span class="fl">14</span>,
  width<span class="op">=</span><span class="fl">0.10</span>, <span class="co"># adds the second row of heatmap describing ciprofloxacin resistance genes</span>
  colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">scale_fill_manual</span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"Ciprofloxacin resistance \n conferring mutation"</span>,
    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#fe9698"</span>,<span class="st">"#ea0c92"</span><span class="op">)</span>,
    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA D87Y"</span>, <span class="st">"gyrA S83L"</span><span class="op">)</span>,
    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA d87y"</span>, <span class="st">"gyrA s83l"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>
    legend.position<span class="op">=</span><span class="st">"bottom"</span>,
    legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,
    legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span><span class="fl">10</span><span class="op">)</span>,
    legend.box <span class="op">=</span> <span class="st">"vertical"</span>,
    legend.margin <span class="op">=</span> <span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>
    fill <span class="op">=</span> <span class="fu">guide_legend</span><span class="op">(</span>nrow<span class="op">=</span><span class="fl">2</span>,byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>

<span class="va">h3</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_cip_genes-1.png" width="70%" style="display: block; margin: auto;"></div>
<p>Next we add continuous data on actual resistance determined by the laboratory
as the minimum inhibitory concentration (MIC) of ciprofloxacin:</p>
<div class="sourceCode" id="cb1522"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># First we add the new coloring scheme:</span>
<span class="va">h4</span> <span class="op">&lt;-</span> <span class="va">h3</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span>

<span class="co"># then we combine the two into a new plot:</span>
<span class="va">h5</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span>
  <span class="va">h4</span>,
  <span class="va">MIC_Cip</span>,
  offset <span class="op">=</span> <span class="fl">16</span>,
  width<span class="op">=</span><span class="fl">0.10</span>,
  colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">scale_fill_continuous</span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"MIC for ciprofloxacin"</span>,
    low <span class="op">=</span> <span class="st">"yellow"</span>,
    high <span class="op">=</span> <span class="st">"red"</span>,
    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.50</span>, <span class="fl">1.00</span><span class="op">)</span>,
    na.value <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>
    fill <span class="op">=</span> <span class="fu">guide_colourbar</span><span class="op">(</span>barwidth <span class="op">=</span> <span class="fl">5</span>, barheight <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>
    legend.position<span class="op">=</span><span class="st">"bottom"</span>,
    legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,
    legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span><span class="fl">10</span><span class="op">)</span>,
    legend.box<span class="op">=</span><span class="st">"vertical"</span>,
    legend.margin<span class="op">=</span><span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">h5</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_cip_MIC-1.png" width="70%" style="display: block; margin: auto;"></div>
<p>We can do the same exercise for a linear tree:</p>
<div class="sourceCode" id="cb1523"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># B: Lineartree:</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">ggtree</span><span class="op">(</span><span class="va">sub_tree2</span><span class="op">)</span> <span class="op">%&lt;+%</span> <span class="va">sample_data</span> <span class="op">+</span>
  <span class="fu">geom_tiplab</span><span class="op">(</span>size <span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="op">+</span> <span class="co"># labels the tips</span>
  <span class="fu">theme_tree2</span><span class="op">(</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">xlab</span><span class="op">(</span><span class="st">"genetic distance"</span><span class="op">)</span><span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.015</span><span class="op">)</span><span class="op">+</span>
 <span class="fu">theme</span><span class="op">(</span>
   legend.position <span class="op">=</span> <span class="st">"none"</span>,
   axis.title.y<span class="op">=</span><span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,
   plot.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span><span class="fl">12</span>, face<span class="op">=</span><span class="st">"bold"</span>,hjust <span class="op">=</span> <span class="fl">0.5</span>, vjust <span class="op">=</span> <span class="op">-</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span>


<span class="co"># First we add gender:</span>

<span class="va">h1</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span>
  <span class="va">p</span>, <span class="va">gender</span>,
  offset <span class="op">=</span> <span class="fl">0.003</span>,
  width<span class="op">=</span><span class="fl">0.1</span>,
  color<span class="op">=</span><span class="st">"black"</span>, 
  colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">scale_fill_manual</span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"Gender"</span>,
    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#00d1b1"</span>, <span class="st">"purple"</span><span class="op">)</span>,
    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span>,
    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
   <span class="fu">theme</span><span class="op">(</span>
     legend.position<span class="op">=</span><span class="st">"bottom"</span>,
     legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,
     legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span><span class="fl">10</span><span class="op">)</span>,
     legend.box<span class="op">=</span><span class="st">"vertical"</span>, legend.margin<span class="op">=</span><span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co"># h1</span>

<span class="co"># Then we add ciprofloxacin after adding another colorscheme layer:</span>

<span class="va">h2</span> <span class="op">&lt;-</span> <span class="va">h1</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span>
<span class="va">h3</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span>
  <span class="va">h2</span>, <span class="va">cipR</span>,
  offset <span class="op">=</span> <span class="fl">0.004</span>,
  width<span class="op">=</span><span class="fl">0.1</span>,
  color<span class="op">=</span><span class="st">"black"</span>,
  colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">scale_fill_manual</span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"Ciprofloxacin resistance \n conferring mutation"</span>,
    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#fe9698"</span>,<span class="st">"#ea0c92"</span><span class="op">)</span>,
    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA D87Y"</span>, <span class="st">"gyrA S83L"</span><span class="op">)</span>,
    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span> <span class="st">"gyrA d87y"</span>, <span class="st">"gyrA s83l"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>
    legend.position<span class="op">=</span><span class="st">"bottom"</span>,
    legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,
    legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span><span class="fl">10</span><span class="op">)</span>,
    legend.box<span class="op">=</span><span class="st">"vertical"</span>,
    legend.margin<span class="op">=</span><span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>
    fill<span class="op">=</span><span class="fu">guide_legend</span><span class="op">(</span>nrow<span class="op">=</span><span class="fl">2</span>,byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="co"># h3</span>

<span class="co"># Then we add the minimum inhibitory concentration determined by the lab (MIC):</span>
<span class="va">h4</span> <span class="op">&lt;-</span> <span class="va">h3</span> <span class="op">+</span> <span class="fu">new_scale_fill</span><span class="op">(</span><span class="op">)</span>
<span class="va">h5</span> <span class="op">&lt;-</span> <span class="fu">gheatmap</span><span class="op">(</span>
  <span class="va">h4</span>, <span class="va">MIC_Cip</span>,
  offset <span class="op">=</span> <span class="fl">0.005</span>,
  width<span class="op">=</span><span class="fl">0.1</span>,
  color<span class="op">=</span><span class="st">"black"</span>,
  colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">scale_fill_continuous</span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"MIC for ciprofloxacin"</span>,
    low <span class="op">=</span> <span class="st">"yellow"</span>,
    high <span class="op">=</span> <span class="st">"red"</span>,
    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.50</span>,<span class="fl">1.00</span><span class="op">)</span>,
    na.value <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span><span class="op">+</span>
   <span class="fu">guides</span><span class="op">(</span>
     fill <span class="op">=</span> <span class="fu">guide_colourbar</span><span class="op">(</span>barwidth <span class="op">=</span> <span class="fl">5</span>, barheight <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
   <span class="fu">theme</span><span class="op">(</span>
     legend.position<span class="op">=</span><span class="st">"bottom"</span>,
     legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size<span class="op">=</span><span class="fl">10</span><span class="op">)</span>,
     legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span><span class="fl">8</span><span class="op">)</span>,
     legend.box<span class="op">=</span><span class="st">"horizontal"</span>, legend.margin<span class="op">=</span><span class="fu">margin</span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>
    shape <span class="op">=</span> <span class="fu">guide_legend</span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">h5</span></code></pre></div>
<div class="inline-figure"><img src="_main_files/figure-html/phylogenetic_trees_sampledata_heatmap_linear-1.png" width="80%" style="display: block; margin: auto;"></div>
<!-- ======================================================= -->
</div>
<div id="resources-31" class="section level2" number="38.5">
<h2>
<span class="header-section-number">38.5</span> Resources<a class="anchor" aria-label="anchor" href="#resources-31"><i class="fas fa-link"></i></a>
</h2>
<p><a href="http://hydrodictyon.eeb.uconn.edu/eebedia/index.php/Ggtree#" class="uri">http://hydrodictyon.eeb.uconn.edu/eebedia/index.php/Ggtree#</a> Clade_Colors
<a href="https://bioconductor.riken.jp/packages/3.2/bioc/vignettes/ggtree/inst/doc/treeManipulation.html" class="uri">https://bioconductor.riken.jp/packages/3.2/bioc/vignettes/ggtree/inst/doc/treeManipulation.html</a>
<a href="https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html" class="uri">https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html</a>
<a href="https://bioconductor.riken.jp/packages/3.8/bioc/vignettes/ggtree/inst/doc/treeManipulation.html" class="uri">https://bioconductor.riken.jp/packages/3.8/bioc/vignettes/ggtree/inst/doc/treeManipulation.html</a></p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="transmission-chains.html"><span class="header-section-number">37</span> Transmission chains</a></div>
<div class="next"><a href="interactive-plots.html"><span class="header-section-number">39</span> Interactive plots</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#phylogenetic-trees"><span class="header-section-number">38</span> Phylogenetic trees</a></li>
<li><a class="nav-link" href="#preparation-30"><span class="header-section-number">38.1</span> Preparation</a></li>
<li>
<a class="nav-link" href="#simple-tree-visualization"><span class="header-section-number">38.2</span> Simple tree visualization</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#different-tree-layouts">Different tree layouts</a></li>
<li><a class="nav-link" href="#simple-tree-with-addition-of-sample-data">Simple tree with addition of sample data</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#manipulation-of-phylogenetic-trees"><span class="header-section-number">38.3</span> Manipulation of phylogenetic trees</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#zooming-in-on-one-part-of-the-tree"><span class="header-section-number">38.3.1</span> Zooming in on one part of the tree:</a></li>
<li><a class="nav-link" href="#collapsing-one-part-of-the-tree"><span class="header-section-number">38.3.2</span> Collapsing one part of the tree:</a></li>
<li><a class="nav-link" href="#subsetting-a-tree">Subsetting a tree</a></li>
<li><a class="nav-link" href="#rotating-nodes-in-a-tree">Rotating nodes in a tree</a></li>
<li><a class="nav-link" href="#example-subtree-with-sample-data-annotation"><span class="header-section-number">38.3.3</span> Example subtree with sample data annotation:</a></li>
</ul>
</li>
<li><a class="nav-link" href="#more-complex-trees-adding-heatmaps-of-sample-data"><span class="header-section-number">38.4</span> More complex trees: adding heatmaps of sample data</a></li>
<li><a class="nav-link" href="#resources-31"><span class="header-section-number">38.5</span> Resources</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>The Epidemiologist R Handbook</strong>" was written by the handbook team. It was last built on 2021-04-05.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>
</html>
